---
  title: "Heatmap Analysis to Genome-wide MHL Matrix in MONOD Project"
author: "Shicheng Guo"
date: "Friday, Feb 18, 2016"
output: html_document
---
  
  Update: Jan-19-2015

```{r}
library("impute")
RawNARemove<-function(data,missratio=0.3){
  threshold<-(missratio)*dim(data)[2]
  NaRaw<-which(apply(data,1,function(x) sum(is.na(x))>threshold))
  zero<-which(apply(data,1,function(x) all(x==0))==T)
  NaRAW<-c(NaRaw,zero)
  if(length(NaRAW)>0){
    dat<-data[-NaRAW,]
  }else{
    dat<-data;
  }
  dat
} 
ci95<-function(x){
  error <- qt(0.975,df=length(x)-1)*sd(x)/sqrt(length(x))
  m<-round(mean(x),2)
  d<-round(mean(x)-error,2)
  u<-round(mean(x)+error,2)
  paste(m, "(",d,"-",u,")",sep="")
}
ttestPValue<-function(data,x1,x2){
  data<-data+matrix(rnorm(nrow(data)*ncol(data),0,0.00000000001),nrow=nrow(data),ncol=ncol(data))
  output<-matrix(NA,dim(data)[1],5)
  for(i in 1:dim(data)[1]){
    out<-data.frame()
    if(all(! any(all(is.na(data[i,x1])),all(is.na(data[i,x2]))),sum(is.na(data[i,]))<0.5*length(data[i,]))){ 
      tmp1<-try(t.test(as.numeric(data[i,x1]),as.numeric(data[i,x2]),paired=F, na.action=na.omit))
      output[i,1]<-tmp1$p.value
      output [i,2]<-mean(as.numeric(data[i,x1]))-mean(as.numeric(data[i,x2]))
      output[i,3]<-"t-test"
      output[i,4]<-mean(as.numeric(data[i,x1]))
      output[i,5]<-mean(as.numeric(data[i,x2]))
    }
  }
  FDR<-p.adjust(output[,1],"fdr")
  out<-data.frame(rownames(data),output,FDR)
  colnames(out)<-c("ID","P-value","Delta(M1-M2)","Test","Mean1","Mean2","FDR")
  out
}
ColNARemove<-function(data,missratio=0.3){
  threshold<-(missratio)*dim(data)[1]
  NaCol<-which(apply(data,2,function(x) sum(is.na(x))>threshold))
  zero<-which(apply(data,2,function(x) all(x==0))==T)
  NaCOL<-c(NaCol,zero)
  if(length(NaCOL)>0){
    data1<-data[,-NaCOL]
  }else{
    data1<-data;
  }
  data1
}
bedwithgap<-function(bed,gap){
  bed<-as.matrix(bed)
  bed[,2]=as.numeric(bed[,2])-gap
  bed[,3]=as.numeric(bed[,3])+gap
  bed<-data.frame(bed)
  bed
}
Rbedtools<-function(functionstring="intersectBed",bed1,bed2,opt.string=""){
  #create temp files
  a.file=tempfile()
  b.file=tempfile()
  out   =tempfile()
  options(scipen =99) # not to use scientific notation when writing out
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  write.table(bed2,file=b.file,quote=F,sep="\t",col.names=F,row.names=F)
  # create the command string and call the command using system()
  command=paste(functionstring,"-a",a.file,"-b",b.file,opt.string,">",out,sep=" ")
  cat(command,"\n")
  try(system(command))
  res=read.table(out,header=F)
  unlink(a.file);unlink(b.file);unlink(out)
  return(res)
}
cor2bed<-function(cor){
  a<-unlist(lapply(strsplit(as.character(cor),split=c(":")),function(x) strsplit(x,"-")))
  bed<-matrix(a,ncol=3,byrow=T)
  return(data.frame(bed))
}
bed2cor<-function(bed){
  cor<-apply(bed,1,function(x){paste(unlist(strsplit(x,"\t"))[1],":",unlist(strsplit(x,"\t"))[2],"-",unlist(strsplit(x,"\t"))[3],sep="")})
  cor<-gsub("[ ]","",cor)
  return(cor)
}
PCAPlot<-function(data,pheno,output,multifigure=T){
  pca <- prcomp(data,center=T,scale = F)  # Here, input file: row is individual and column is variable
  outputfile=paste(output,".pdf",sep="")
  pdf(outputfile)
  if(multifigure){
    par(mfrow=c(2,2),mar=c(4,4,4,4))  
  }
  plot((pca$sdev[1:10])^2,type="o",xaxt="n",ylab="Variances",xlab="Principle Components",col="red",lwd=2)
  axis(1,at=0:10,labels=paste("PC",0:10,sep=""))
  var<-c()
  for(i in 1:length(pca$sdev)){var[i]<-sum((pca$sdev[1:i])^2)/sum((pca$sdev)^2)}
  plot(var,ylab="total variance",xlab="number of principle components",lwd=2,type="l")
  abline(h=0.8,col="grey",lty=2)
  abline(v=which(var>0.8)[1],col="grey",lty=2)
  scores <- data.frame(pheno, pca$x[,1:3])
  col = as.numeric(as.factor(pheno))
  plot(x=scores$PC1,y=scores$PC2, xlim=c(min(scores$PC1),max(scores$PC1)),ylim=c(min(scores$PC2),max(scores$PC2)),type="n",xlab="PC1",ylab="PC2")
  for(i in 1:length(scores$PC1)){
    points(scores$PC1[i],scores$PC2[i],pch=as.numeric(as.factor(pheno))[i],col=col[i],cex=0.8,lwd=2)
  }
  legend("bottomright",legend=names(table(pheno)),pch=1:length(table(pheno)),col=1:length(table(pheno)),bty="n")
  plot(x=scores$PC1,y=scores$PC3, xlim=c(min(scores$PC1),max(scores$PC1)),ylim=c(min(scores$PC3),max(scores$PC3)),type="n",xlab="PC1",ylab="PC3")
  for(i in 1:length(scores$PC1)){
    points(scores$PC1[i],scores$PC3[i],pch=as.numeric(as.factor(pheno))[i],col=col[i],cex=0.9,lwd=2)
  }
  legend("bottomright",legend=names(table(pheno)),pch=1:length(table(pheno)),col=1:length(table(pheno)),bty="n")
  dev.off()
}
gsi<-function(data){
  group=names(table(colnames(data)))
  index=colnames(data)
  gsi<-c()
  gmaxgroup<-c()
  for(i in 1:nrow(data)){
    gsit<-0
    gmax<-names(which.max(tapply(as.numeric(data[i,]),index,mean)))
    for(j in 1:length(group)){
      tmp<-(1-10^(mean(data[i,][which(index==group[j])]))/10^(mean(data[i,][which(index==gmax)])))/(length(group)-1)
      gsit<-gsit+tmp
    }
    gmaxgroup<-c(gmaxgroup,gmax)
    gsi<-c(gsi,gsit)
    print(c(gmax,gsit))
  }
  rlt=data.frame(region=rownames(data),group=gmaxgroup,GSI=gsi)
  return(rlt)
}
```


```{r}
setwd("/home/shg047/monod/dec")
infile="WGBS_methHap_load_matrix_20Oct2015.txt";
file1<-read.table(infile,head=T,sep="\t",row.names=1,as.is=T,check.names=F)

# miss value detection and imputation
library("impute")
f2<-RawNARemove(file1,missratio=0.3)
f2<-impute.knn(data.matrix(f2))$data

library("preprocessCore")
f2.t1<-normalize.quantiles(f2[,13:58])
library("sva")
batch=c(rep(1,10),rep(2,36))
f2.t2<-ComBat(f2.t1, batch, mod=NULL, par.prior = TRUE,prior.plots = FALSE)
f2[,13:58]<-f2.t2

# re-plot the cluster by sequencing sample Tag
colnames(f2)<-colnames(file1)
f3 <- t(na.omit(f2)) # listwise deletion of missing
histFfile=paste("Figure.Cluster.mhl.Dendrogram.combat.seqTag.ward.D.plot",infile,"pdf",sep=".")
d <- dist(f3, method = "euclidean") # distance matrix
fit <- hclust(d, method="ward.D") 
pdf(histFfile)
plot(fit,cex=0.5,hang=-1) # display dendogram
dev.off()
histFfile=paste("Figure.Cluster.mhl.Dendrogram.combat.seqTag.plot.average",infile,"pdf",sep=".")
d <- dist(f3, method = "euclidean") # distance matrix
fit <- hclust(d, method="average") 
pdf(histFfile)
plot(fit,cex=0.5,hang=-1) # display dendogram
dev.off()

# re-assign colnames
colnames(f2) 
colnames(f2)<-gsub("_","-",colnames(f2))
colname2<-unlist(lapply(colnames(f2),function(x) unlist(strsplit(x,"[.]"))[1]))
colname2
colnames(f2)<-colname2
# be sure all the sample information has been stored in the following database
saminfo2<-read.table("/home/shg047/monod/phase2/newsaminfo.txt",head=T,sep="\t",as.is=T)
saminfo2<-saminfo2[match(colname2,saminfo2[,1]),]
saminfo2
colnames(f2)<-saminfo2[,2]
colnames(f2)

# check the dendrogram after missing, quantile
f3 <- t(na.omit(f2)) # listwise deletion of missing
# f3 <- scale(f3) # standardize variables
sum(is.na(file1))/(nrow(file1)*ncol(file1))
sum(is.na(f2))/(nrow(f2)*ncol(f2))
histFfile=paste("Figure.hist.combat",infile,"pdf",sep=".")
pdf(histFfile)
boxplot(f2,outline=F,horizontal=T,notch=F,las=1,cex.axis=0.65,col="blue")
dev.off()
histFfile=paste("Figure.Cluster.mhl.Dendrogram.combat.ward.D.plot",infile,"pdf",sep=".")
d <- dist(f3, method = "euclidean") # distance matrix
fit <- hclust(d, method="ward.D") 
pdf(histFfile)
plot(fit,cex=0.5,hang=-1) # display dendogram
dev.off()
histFfile=paste("Figure.Cluster.mhl.Dendrogram.combat.plot.average",infile,"pdf",sep=".")
d <- dist(f3, method = "euclidean") # distance matrix
fit <- hclust(d, method="average") 
pdf(histFfile)
plot(fit,cex=0.5,hang=-1) # display dendogram
dev.off()


# tissue specific hml
gsi_1<-gsi(f2[,13:58])
gsi_2<-gsi(f2)

gsi1<-subset(gsi_1,GSI>0.6)
gsi2<-subset(gsi_2,GSI>0.6)

dim(gsi1)
dim(gsi2)

write.table(gsi1,file="Table.GSI.WGBS.Remove.H1.WBC.rlt.txt",col.names=T,row.names=F,quote=F,sep="\t")
write.table(gsi2,file="Table.GSI.WGBS.with.H1.WBC.rlt.txt",col.names=T,row.names=F,quote=F,sep="\t")
```

```{r}
setwd("C:\\Users\\shicheng\\Dropbox\\Project\\methylation\\monod\\result\\Figure 3")
load("wgbs.mhl.heatmap.RData")
library("gplots")
library("RColorBrewer")
library("grDevices")
mydata<-f2[which(gsi_2[,3]>0.6),]
save(mydata,file="GWBS.mhl.heatmap.RData")
hclustfunc <- function(x) hclust(x, method="ward.D")
distfunc <- function(x) dist(x, method="euclidean")
# perform clustering on rows and columns
cl.row <- hclustfunc(distfunc(mydata))
cl.col <- hclustfunc(distfunc(t(mydata)))
# extract cluster assignments; i.e. k=8 (rows) k=5 (columns), ncut should smaller than total color number. "Set1" of brewer.pal <12
ncut=6  
gr.row <- cutree(cl.row, ncut)
# gr.col <- cutree(cl.col, 5)
require(RColorBrewer)
col1 <- brewer.pal(ncut, "Set1")[gr.row[cl.row$order]]     # the maximum of brewer.pal is 12
#col2 <- brewer.pal(5, "Pastel1")

c25 <- c("dodgerblue2","#E31A1C", # red
         "green4",
         "#6A3D9A", # purple
         "#FF7F00", # orange
         "black","gold1",
         "skyblue2","#FB9A99", # lt pink
         "palegreen2",
         "#CAB2D6", # lt purple
         "#FDBF6F", # lt orange
         "gray70", "khaki2",
         "maroon","orchid1","deeppink1","blue1","steelblue4",
         "darkturquoise","green1","yellow4","yellow3",
         "darkorange4","brown")

coll<-gr.row[as.numeric(as.factor(cl.row$labels))]
col2<-c25[as.numeric(as.factor(cl.col$labels))]
library("grDevices")
library("gplots")
mydata[mydata<0]<-0
mydata[mydata>1]<-1
col=colorRampPalette(c("blue", "yellow"))(20) 
save(mydata,file="wgbs.mhl.heatmap.RData")
pdf("heatmap.wgbs.mhl.gsi0.6.with.H1.wbc.cancer.pdf")
heatmap.2(as.matrix((mydata)),hclustfun=hclustfunc, distfun=distfunc,
          ColSideColors=col2,
          Colv=T,Rowv=T,key=T,
          col=col,
          cexCol=0.65,
          keysize=1,
          labRow=NA,
          margins=c(5,10),
          trace="none",
          srtCol=45,
          density.info="none")
legend=unique(data.frame(col2,cl.col$labels))
legend(x=0.85,y=0.8,legend=legend[,2],col=as.character(legend[,1]),pch=15,cex = 0.5)
dev.off()


```