<!DOCTYPE html>
<!-- saved from url=(0097)http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes -->
<html lang="en" dir="ltr" class="client-js"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>Dinh/Dinh 2014/NOTES/2014-9-15 - ZhangLabWiki</title>
<meta charset="UTF-8">
<meta name="generator" content="MediaWiki 1.18.2">
<meta name="robots" content="noindex,follow">
<link rel="alternate" type="application/x-wiki" title="Edit" href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&action=edit">
<link rel="edit" title="Edit" href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&action=edit">
<link rel="shortcut icon" href="http://genome-tech.ucsd.edu/favicon.ico">
<link rel="search" type="application/opensearchdescription+xml" href="http://genome-tech.ucsd.edu/LabNotes/opensearch_desc.php" title="ZhangLabWiki (en)">
<link rel="EditURI" type="application/rsd+xml" href="http://genome-tech.ucsd.edu/LabNotes/api.php?action=rsd">
<link rel="copyright" href="http://www.gnu.org/copyleft/fdl.html">
<link rel="alternate" type="application/atom+xml" title="ZhangLabWiki Atom feed" href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Special:RecentChanges&feed=atom">
<link rel="stylesheet" href="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/load.php">
<style type="text/css" media="all">.mw-collapsible-toggle{float:right} li .mw-collapsible-toggle{float:none} .mw-collapsible-toggle-li{list-style:none}

/* cache key: wikidb-LabNotes:resourceloader:filter:minify-css:4:4250852ed2349a0d4d0fc6509a3e7d4c */
</style><style type="text/css" media="all">.js-messagebox{margin:1em 5%;padding:0.5em 2.5%;border:1px solid #ccc;background-color:#fcfcfc;font-size:0.8em}.js-messagebox .js-messagebox-group{margin:1px;padding:0.5em 2.5%;border-bottom:1px solid #ddd}.js-messagebox .js-messagebox-group:last-child{border-bottom:thin none transparent}

/* cache key: wikidb-LabNotes:resourceloader:filter:minify-css:4:8b08bdc91c52a9ffba396dccfb5b473c */
</style><meta name="ResourceLoaderDynamicStyles" content="">
<style>a:lang(ar),a:lang(ckb),a:lang(fa),a:lang(kk-arab),a:lang(mzn),a:lang(ps),a:lang(ur){text-decoration:none}a.new,#quickbar a.new{color:#ba0000}

/* cache key: wikidb-LabNotes:resourceloader:filter:minify-css:4:c88e2bcd56513749bec09a7e29cb3ffa */
</style>
<script src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/load(1).php"></script><script src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/load(2).php"></script>
<script>if(window.mw){
	mw.config.set({"wgCanonicalNamespace": "", "wgCanonicalSpecialPageName": false, "wgNamespaceNumber": 0, "wgPageName": "Dinh/Dinh_2014/NOTES/2014-9-15", "wgTitle": "Dinh/Dinh 2014/NOTES/2014-9-15", "wgCurRevisionId": 62846, "wgArticleId": 19089, "wgIsArticle": true, "wgAction": "view", "wgUserName": "Shicheng", "wgUserGroups": ["*", "user", "autoconfirmed"], "wgCategories": [], "wgBreakFrames": false, "wgRestrictionEdit": [], "wgRestrictionMove": []});
}
</script><script>if(window.mw){
	mw.loader.load(["mediawiki.page.startup"]);
}
</script><script type="text/javascript" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/load(3).php"></script>
</head>
<body class="mediawiki ltr sitedir-ltr ns-0 ns-subject page-Dinh_Dinh_2014_NOTES_2014-9-15 action-view skin-monobook">
<div id="globalWrapper">
<div id="column-content"><div id="content"><div id="mw-js-message" class="js-messagebox" style="display: none;"></div>
	<a id="top"></a>
	
	<h1 id="firstHeading" class="firstHeading">Dinh/Dinh 2014/NOTES/2014-9-15</h1>
	<div id="bodyContent">
		<div id="siteSub">From ZhangLabWiki</div>
		<div id="contentSub"></div>
		<div id="jump-to-nav">Jump to: <a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#column-one">navigation</a>, <a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#searchInput">search</a></div>
		<!-- start content -->
<div lang="en" dir="ltr" class="mw-content-ltr"><table id="toc" class="toc"><tbody><tr><td><div id="toctitle"><h2>Contents</h2><span class="toctoggle">&nbsp;[<a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#" class="internal" id="togglelink">hide</a>]&nbsp;</span></div>
<ul>
<li class="toclevel-1 tocsection-1"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#MethylPurify"><span class="tocnumber">1</span> <span class="toctext">MethylPurify</span></a>
<ul>
<li class="toclevel-2 tocsection-2"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#Test_run_with_WGBS_data_and_BSMAP_aligner"><span class="tocnumber">1.1</span> <span class="toctext">Test run with WGBS data and BSMAP aligner</span></a></li>
<li class="toclevel-2 tocsection-3"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#Testing_with_MONOD_RRBS_bam_files_from_BisReadMapper"><span class="tocnumber">1.2</span> <span class="toctext">Testing with MONOD RRBS bam files from BisReadMapper</span></a></li>
<li class="toclevel-2 tocsection-4"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#MethylPurify_run_on_merged_duplicates_tumor_samples"><span class="tocnumber">1.3</span> <span class="toctext">MethylPurify run on merged duplicates tumor samples</span></a></li>
<li class="toclevel-2 tocsection-5"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#Overlap_of_informative_CGIs_with_blood_UMR.2FLMRs"><span class="tocnumber">1.4</span> <span class="toctext">Overlap of informative CGIs with blood UMR/LMRs</span></a></li>
<li class="toclevel-2 tocsection-6"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#Methylation_status_of_informative_bins_in_normal_lung.2C_colon.2C_and_pancreas"><span class="tocnumber">1.5</span> <span class="toctext">Methylation status of informative bins in normal lung, colon, and pancreas</span></a>
<ul>
<li class="toclevel-3 tocsection-7"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#Download_ENCODE_data"><span class="tocnumber">1.5.1</span> <span class="toctext">Download ENCODE data</span></a></li>
<li class="toclevel-3 tocsection-8"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#Generate_methylation_matrices_over_Informative_Bins"><span class="tocnumber">1.5.2</span> <span class="toctext">Generate methylation matrices over Informative Bins</span></a></li>
<li class="toclevel-3 tocsection-9"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#.22Normal.22_reference_bins"><span class="tocnumber">1.5.3</span> <span class="toctext">"Normal" reference bins</span></a></li>
</ul>
</li>
<li class="toclevel-2 tocsection-10"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#Divide_haplotypes_into_clusters_in_solid_tumor_samples"><span class="tocnumber">1.6</span> <span class="toctext">Divide haplotypes into clusters in solid tumor samples</span></a></li>
<li class="toclevel-2 tocsection-11"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&printable=yes#Discussion"><span class="tocnumber">1.7</span> <span class="toctext">Discussion</span></a></li>
</ul>
</li>
</ul>
</td></tr></tbody></table>
<h1> <span class="mw-headline" id="MethylPurify"> MethylPurify </span></h1>
<ul><li> Tumor deconvolution paper: <a rel="nofollow" class="external free" href="http://genomebiology.com/2014/15/8/419">http://genomebiology.com/2014/15/8/419</a>
</li><li> Python package: <a rel="nofollow" class="external free" href="https://pypi.python.org/pypi/MethylPurify/2.0-20140819">https://pypi.python.org/pypi/MethylPurify/2.0-20140819</a>
</li><li> Installed MethylPurify on genome-miner. The scripts dont take much memory &amp; finish relatively quickly.
</li></ul>
<h2> <span class="mw-headline" id="Test_run_with_WGBS_data_and_BSMAP_aligner"> Test run with WGBS data and BSMAP aligner </span></h2>
<ul><li> I tried running MethylPurify on our RRBS MONOD bam files, but could never get the program to complete running/obtain informative bins and mixing ratio.
</li><li> Two major issues with MethylPurify: requires enough reads to get minimum depth of coverage over informative bins and may need to realign data with BSMAP.
</li><li> Downloaded WGBS sequencing data from the paper: SRR1232303 (BioSample: SAMN02725471; lung adenocarcinoma tissue, male), containing ~20 million PE reads x 90 bp each.
</li><li> Downsampled to get ~25% of only reads 1 to get 5,077,371 x 90 bp single end reads to perform testing.
</li><li> Performed mapping with BSMAP as that aligner is listed by MethylPurify to get the necessary bam input file.
</li><li> BSMAP mapping command:
</li></ul>
<pre> /bsmap-2.74/bsmap -u -s 12 -v 0.04 -p 4 -a downsampled.SRR1232303_1.fastq -d hg19.fa -o downsampled.SRR1232303_1.fastq.sam
</pre>
<ul><li> BSMAP took 36,221 seconds (10.06 hours) to complete mapping for ~5 million SE x 90 bp reads with 4 CPU threads.
</li><li> Performed deconvolution with MethylPurify:
</li></ul>
<pre> python2.7 /home/ddiep/Downloads/MethylPurify-2.0-20140819/methylpurify/bin/MethylPurify -f downsampled.SRR1232303_1.fastq.bam 
      -g hg19.fa -i /home/ddiep/Downloads/MethylPurify-2.0/methylpurify/db/CGI_hg19_slop1000.bed -c 10 -s 50 -b 300
</pre>
<ul><li> MethylPurify outputs:
</li></ul>
<pre> ...
 The predicted mixing ratio is:  0.265
 The number of informative bins is:  29.0
 The number of informative bins is less than 400, too less for the prediction of mixing ratio, continue
 INFO  @ Mon, 15 Sep 2014 09:33:27: Running get_methyl_profile:
 MixingRatio:0.265
</pre>
<ul><li> CGI_hg19_slop1000.bed was only found in the MethylPurify-2.0.tar.gz package, but not the latest package. I couldn't find the genome.sh script that was necessary to generate this, so I need to figure out how to generate this file in the future because I am not 100% confident this is the appropriate file to use.
</li></ul>
<h2> <span class="mw-headline" id="Testing_with_MONOD_RRBS_bam_files_from_BisReadMapper"> Testing with MONOD RRBS bam files from BisReadMapper </span></h2>
<ul><li> Since WGBS and RRBS may cover different CGIs, the CGI selected regions by MethylPurify have this depth of coverage:
</li></ul>
<pre> samtools depth XXX.CGI.sorted.bam  |  awk '{sum+=$3} END { print "Sum = ", sum, " Average = ",sum/NR}' 
 &gt;&gt; Sum =  41941637  Average =  2.30644
</pre>
<ul><li> Our RRBS data have even better coverage at CGIs, the CGI regions selected by MethylPurify for PC-T-7_2 have this depth of coverage:
</li></ul>
<pre> samtools depth PC-T-7_2.merged.CGI.sorted.bam  |  awk '{sum+=$3} END { print "Sum = ", sum, " Average = ",sum/NR}'
 &gt;&gt; Sum =  393749988  Average =  27.9027
</pre>
<ul><li> This could means that MethylPurify failed when I ran on our RRBS bam files because it couldn't understand the alignments from bisReadMapper, but requires the BSMAP alignment format (may have special instructions for MethylPurify). I will figure out the differences in BSMAP and BisReadMapper bam files.
</li><li> BSMAP sam file:
</li></ul>
<pre>SRR1232303.200105       0       chr1    30557958        255     90M     *       0       0       ATTCCTTTATTTTGGTGGTTGGGGTGTTTTAGGGTTTTTTTTTTAATTAATTTTTTTTTCGTTGTTTTTTTTTTTTTGTTTTATGTTTTT      @CCFFFFFFHHHHJICGHHIJJJJ:DFHIJHJJJHJJJJJJHFDCDDEECEEEDDDDDDDDDDDDDDDDDDDDDDDD38ACD3:(:ADDD      NM:i:3  ZS:Z:++
SRR1232303.200112       16      chr1    30559194        255     90M     *       0       0       ATCCCTTCCCTTTCACAACCTACTCCGAAAACCGCTAACCTAAAAATACAGCAATCTCTACTATCCTAAAAAAACAAAATCCCAAGGAAT      DDDDDDDDBDDEDDDDDDDDCACAFFFHHGIJJIJIHHJIHJJJIIGJJJJJIJGIIJJIJJIHJIHJJJJJJJJJJHHHHHFFFFFCCC      NM:i:2  ZS:Z:-+
</pre>
<ul><li> BSMAP have extra ZS:Z field that's just the strand information. MethylPurify would fail similarly as for BisReadMapper bam files, after I removed the ZS:Z field from the BSMAP bam file. I may just need to generate the ZS:Z field to get MethylPurify to work on BisReadMapper bam files.
</li><li> Encoding for ZS:Z field is as follows:
</li></ul>
<pre>       ++: forward strand of Watson of reference (BSW)
       +-: reverse strand of Watson of reference (BSWC)
       -+: forward strand of Crick of reference (BSC)  
       --: reverse strand of Crick of reference (BSCC)
</pre>
<ul><li> Make MethylPurify think that BisReadMapper bam files are from BSMAP: Add in NM:i:0 and ZS:Z:-+ (for Crick) and ZS:Z:++ (for Watson).
</li></ul>
<pre> samtools view $f | awk '{if($2 == 0) print $0"\tNM:i:0\tZS:Z:++"; if($2 == 16) print $0"\tNM:i:0\tZS:Z:-+"}' | samtools view -bSt /media/2TB_storeA/BisRef/Hg19_plusLamba/hg19_lambda.fa.fai - &gt; mod.$f
</pre>
<ul><li> Run MethylPurify:
</li></ul>
<pre> python2.7 /home/ddiep/Downloads/MethylPurify-2.0-20140819/methylpurify/bin/MethylPurify -f ../mod.PC-T-7_2.merged.bam -g hg19.fa -i /home/ddiep/Downloads/MethylPurify-2.0/methylpurify/db/CGI_hg19_slop1000.bed -c 10 -s 50 -b 300
</pre>
<ul><li> MethylPurify outputs:
</li></ul>
<pre> ...
 The predicted mixing ratio is:  0.175
 The number of informative bins is:  304.0
 The number of informative bins is less than 400, too less for the prediction of mixing ratio, continue
 INFO  @ Mon, 15 Sep 2014 13:09:12: Running get_methyl_profile:
 MixingRatio:0.175
</pre>
<ul><li> Results from MethylPurify:
</li></ul>
<pre> <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:MixRatio.MethylPurify_test1.png" class="image"><img alt="MixRatio.MethylPurify test1.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/MixRatio.MethylPurify_test1.png" width="600" height="167"></a>
 <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:InformativeBins.MethylPurify_test1.png" class="image"><img alt="InformativeBins.MethylPurify test1.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/InformativeBins.MethylPurify_test1.png" width="600" height="167"></a>
</pre>
<h2> <span class="mw-headline" id="MethylPurify_run_on_merged_duplicates_tumor_samples"> MethylPurify run on merged duplicates tumor samples</span></h2>
<ul><li> Results from MethylPurify:
</li></ul>
<pre> <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:MixRatio.MethylPurify_testMergedTumors.png" class="image"><img alt="MixRatio.MethylPurify testMergedTumors.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/MixRatio.MethylPurify_testMergedTumors.png" width="300" height="179"></a>
 <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:InformativeBins.MethylPurify_testMergedTumors.png" class="image"><img alt="InformativeBins.MethylPurify testMergedTumors.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/InformativeBins.MethylPurify_testMergedTumors.png" width="300" height="179"></a>
</pre>
<ul><li> I merged all of the informative bins for colon cancer, lung cancer, and pancreatic cancers together into one BED file.
</li></ul>
<pre> <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Colon_cancer_methylpurify_informativeCGIs_merged.txt" title="File:Colon cancer methylpurify informativeCGIs merged.txt">File:Colon cancer methylpurify informativeCGIs merged.txt</a>
 <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Lung_cancer_methylpurify_informativeCGIs_merged.txt" title="File:Lung cancer methylpurify informativeCGIs merged.txt">File:Lung cancer methylpurify informativeCGIs merged.txt</a>
 <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Pancreatic_cancer_methylpurify_informativeCGIs_merged.txt" title="File:Pancreatic cancer methylpurify informativeCGIs merged.txt">File:Pancreatic cancer methylpurify informativeCGIs merged.txt</a>
</pre>
<ul><li> I uploaded the BED files to GREAT &amp; used default settings to get the biological processes that are enriched. This may be because the regions are all CGIs, so not certain whether the signal is real or not.
</li><li> Results from GREAT:
</li></ul>
<pre> <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Colon_cancer_informative_CGIs_biologicalprocess.png" class="image"><img alt="Colon cancer informative CGIs biologicalprocess.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/Colon_cancer_informative_CGIs_biologicalprocess.png" width="500" height="210"></a>
 <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Lung_cancer_informative_CGIs_biologicalprocess.png" class="image"><img alt="Lung cancer informative CGIs biologicalprocess.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/Lung_cancer_informative_CGIs_biologicalprocess.png" width="500" height="222"></a>
 <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Pancreatic_cancer_informative_CGIs_biologicalprocess.png" class="image"><img alt="Pancreatic cancer informative CGIs biologicalprocess.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/Pancreatic_cancer_informative_CGIs_biologicalprocess.png" width="500" height="232"></a>
</pre>
<h2> <span class="mw-headline" id="Overlap_of_informative_CGIs_with_blood_UMR.2FLMRs"> Overlap of informative CGIs with blood UMR/LMRs </span></h2>
<ul><li> Blood UMRs and LMRs were identified using MethylSeekR
</li><li> I used bedtools intersect and required that 50% of the informative bins be overlapped by UMRs/LMRs
</li><li> Results:
</li></ul>
<table class="wikitable">
<tbody><tr>
<td align="center" style="background:#f0f0f0;">'
</td>
<td align="center" style="background:#f0f0f0;"><b>colon</b>
</td>
<td align="center" style="background:#f0f0f0;"><b>lung</b>
</td>
<td align="center" style="background:#f0f0f0;"><b>pancreatic</b>
</td></tr>
<tr>
<td> Number of informative CGIs</td>
<td>2852</td>
<td>1248</td>
<td>1798
</td></tr>
<tr>
<td> overlap with UMR</td>
<td>2274</td>
<td>1144</td>
<td>1699
</td></tr>
<tr>
<td> overlap with LMR</td>
<td>12</td>
<td>14</td>
<td>14
</td></tr>
<tr>
<td>&nbsp;%overlap with UMR</td>
<td>79.73%</td>
<td>91.67%</td>
<td>94.49%
</td></tr>
<tr>
<td>&nbsp;%overlap with LMR</td>
<td>0.42%</td>
<td>1.12%</td>
<td>0.78%
</td></tr>
<tr>
<td>&nbsp;% no overlap</td>
<td>19.85%</td>
<td>7.21%</td>
<td>4.73%
</td></tr></tbody></table>
<ul><li> Majority of informative CGIs are in UMRs.
</li><li> Since only a fraction of informative CGIs are overlapping with LMRs, I looked for region/functional enrichment using GREAT with the CGIs overlapping LMRs as test data and all the informative CGIs as background. No term was enriched.
</li><li> Since most of the informative CGIs are overlapping with UMRs, no term was enriched if I used all informative CGIs as background. So I used the whole genome was background. The GREAT results show developmental associated terms enriched in biological processes.
</li><li> GREAT results:
</li></ul>
<pre> <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Colon_cancer_informative_CGIs_UMRs_biologicalprocess.png" class="image"><img alt="Colon cancer informative CGIs UMRs biologicalprocess.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/Colon_cancer_informative_CGIs_UMRs_biologicalprocess.png" width="500" height="209"></a>
 <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Lung_cancer_informative_CGIs_UMRs_biologicalprocess.png" class="image"><img alt="Lung cancer informative CGIs UMRs biologicalprocess.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/Lung_cancer_informative_CGIs_UMRs_biologicalprocess.png" width="500" height="200"></a>
 <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Pancreatic_cancer_informative_CGIs_UMRs_biologicalprocess.png" class="image"><img alt="Pancreatic cancer informative CGIs UMRs biologicalprocess.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/Pancreatic_cancer_informative_CGIs_UMRs_biologicalprocess.png" width="500" height="230"></a>
</pre>
<h2> <span class="mw-headline" id="Methylation_status_of_informative_bins_in_normal_lung.2C_colon.2C_and_pancreas"> Methylation status of informative bins in normal lung, colon, and pancreas </span></h2>
<ul><li> Define "informative bins"&nbsp;: regions identified using MethylPurify which can indicate the level of heterogeneity in tumor samples AND these regions are also overlapping with blood UMRs (meaning that they are unmethylated in blood tissues).
</li><li> Strategy: (1) Download ENCODE RRBS data for normal tissues (2) Generate a methylation matrix of normal tissues at these bins
</li></ul>
<h3> <span class="mw-headline" id="Download_ENCODE_data"> Download ENCODE data </span></h3>
<ul><li> GEO experiment id: GSE27584
</li><li> Downloaded the tar of bed file.
</li><li> Normal tissues (33 normal samples, 66 libraries) and Cancer tissues (19 cancer samples, 49 libraries)
</li></ul>
<table class="wikitable">
<tbody><tr>
<td align="center" style="background:#f0f0f0;"><b>GSM</b>
</td>
<td align="center" style="background:#f0f0f0;"><b>Sample ID</b>
</td>
<td align="center" style="background:#f0f0f0;"><b>Cell karyotype</b>
</td></tr>
<tr>
<td> GSM683779</td>
<td>BJ</td>
<td>normal
</td></tr>
<tr>
<td> GSM683770</td>
<td>H1-hESC</td>
<td>normal
</td></tr>
<tr>
<td> GSM683927</td>
<td>GM12878</td>
<td>relatively normal
</td></tr>
<tr>
<td> GSM683803</td>
<td>BC_Uterus_BN0765</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683911</td>
<td>BC_Uterus_BN0765</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683781</td>
<td>BC_Testis_N30</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683851</td>
<td>HRCEpiC</td>
<td>normal
</td></tr>
<tr>
<td> GSM683789</td>
<td>BC_Skeletal_Muscle_01-11002</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683941</td>
<td>HSMM</td>
<td>normal
</td></tr>
<tr>
<td> GSM683876</td>
<td>BC_Leukocyte_UHN00204</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683865</td>
<td>BC_Skin_01-11002</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683847</td>
<td>Fibrobl</td>
<td>normal
</td></tr>
<tr>
<td> GSM683841</td>
<td>GM12878-XiMat</td>
<td>relatively normal
</td></tr>
<tr>
<td> GSM683879</td>
<td>H1-hESC</td>
<td>normal
</td></tr>
<tr>
<td> GSM683782</td>
<td>BC_Liver_01-11002</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683883</td>
<td>BC_Adrenal_Gland_H12803N</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683811</td>
<td>BC_Skeletal_Muscle_01-11002</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683834</td>
<td>BC_Breast_02-03015</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683801</td>
<td>HSMMtube</td>
<td>normal
</td></tr>
<tr>
<td> GSM683760</td>
<td>PrEC</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683838</td>
<td>PrEC</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683763</td>
<td>BJ</td>
<td>normal
</td></tr>
<tr>
<td> GSM683821</td>
<td>BC_Lung_01-11002</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683878</td>
<td>BC_Kidney_01-11002</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683796</td>
<td>BC_Placenta_UHN00189</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683765</td>
<td>Astrocy</td>
<td>normal
</td></tr>
<tr>
<td> GSM683786</td>
<td>BC_Adrenal_Gland_H12803N</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683798</td>
<td>BC_Brain_H11058N</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683939</td>
<td>HTR8svn</td>
<td>normal
</td></tr>
<tr>
<td> GSM683901</td>
<td>HRCEpiC</td>
<td>normal
</td></tr>
<tr>
<td> GSM683882</td>
<td>Astrocy</td>
<td>normal
</td></tr>
<tr>
<td> GSM683908</td>
<td>BC_Brain_H11058N</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683892</td>
<td>BC_Skin_01-11002</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683867</td>
<td>HTR8svn</td>
<td>normal
</td></tr>
<tr>
<td> GSM683819</td>
<td>BC_Pancreas_H12817N</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683906</td>
<td>GM12878</td>
<td>relatively normal
</td></tr>
<tr>
<td> GSM683928</td>
<td>Osteobl</td>
<td>normal
</td></tr>
<tr>
<td> GSM683947</td>
<td>BC_Stomach_01-11002</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683820</td>
<td>BC_Pericardium_H12529N</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683846</td>
<td>BC_Left_Ventricle_N41</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683899</td>
<td>BC_Pericardium_H12529N</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683848</td>
<td>BC_Pancreas_H12817N</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683774</td>
<td>GM12878-XiMat</td>
<td>relatively normal
</td></tr>
<tr>
<td> GSM683914</td>
<td>BC_Liver_01-11002</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683759</td>
<td>BC_Leukocyte_UHN00204</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683870</td>
<td>BC_Kidney_01-11002</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683877</td>
<td>BC_Stomach_01-11002</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683916</td>
<td>SAEC</td>
<td>normal
</td></tr>
<tr>
<td> GSM683766</td>
<td>Fibrobl</td>
<td>normal
</td></tr>
<tr>
<td> GSM683943</td>
<td>HRE</td>
<td>normal
</td></tr>
<tr>
<td> GSM683850</td>
<td>BC_Testis_N30</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683764</td>
<td>Hepatocytes</td>
<td>normal
</td></tr>
<tr>
<td> GSM683761</td>
<td>BC_Breast_02-03015</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683895</td>
<td>BC_Left_Ventricle_N41</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683931</td>
<td>BC_Skeletal_Muscle_H12817N</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683831</td>
<td>HRE</td>
<td>normal
</td></tr>
<tr>
<td> GSM683893</td>
<td>BC_Placenta_UHN00189</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683868</td>
<td>BC_Skeletal_Muscle_H12817N</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683907</td>
<td>HSMM</td>
<td>normal
</td></tr>
<tr>
<td> GSM683945</td>
<td>HSMMtube</td>
<td>normal
</td></tr>
<tr>
<td> GSM683818</td>
<td>HSMM</td>
<td>normal
</td></tr>
<tr>
<td> GSM683872</td>
<td>Hepatocytes</td>
<td>normal
</td></tr>
<tr>
<td> GSM683833</td>
<td>SAEC</td>
<td>normal
</td></tr>
<tr>
<td> GSM683940</td>
<td>BC_Lung_01-11002</td>
<td>Normal
</td></tr>
<tr>
<td> GSM683799</td>
<td>HSMMtube</td>
<td>normal
</td></tr>
<tr>
<td> GSM683881</td>
<td>Osteobl</td>
<td>normal
</td></tr></tbody></table>
<table class="wikitable">
<tbody><tr>
<td align="center" style="background:#f0f0f0;"><b>GSM</b>
</td>
<td align="center" style="background:#f0f0f0;"><b>Sample ID</b>
</td>
<td align="center" style="background:#f0f0f0;"><b>Cell karyotype</b>
</td></tr>
<tr>
<td> GSM683792</td>
<td> Jurkat</td>
<td>cancer
</td></tr>
<tr>
<td> GSM980577</td>
<td> ovcar-3</td>
<td>cancer
</td></tr>
<tr>
<td> GSM720353</td>
<td> MCF-7</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683806</td>
<td> Jurkat</td>
<td>cancer
</td></tr>
<tr>
<td> GSM720354</td>
<td> HEK293</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683769</td>
<td> HEK293</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683857</td>
<td> NB4</td>
<td>cancer
</td></tr>
<tr>
<td> GSM980582</td>
<td> Caco-2</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683946</td>
<td> LNCaP</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683863</td>
<td> LNCaP</td>
<td>cancer
</td></tr>
<tr>
<td> GSM980580</td>
<td> A549</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683856</td>
<td> K562</td>
<td>cancer
</td></tr>
<tr>
<td> GSM980575</td>
<td> MCF-7</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683884</td>
<td> SK-N-MC</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683788</td>
<td> HCT-116</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683828</td>
<td> HeLa-S3</td>
<td>cancer
</td></tr>
<tr>
<td> GSM1048458</td>
<td> PANC-1</td>
<td>cancer
</td></tr>
<tr>
<td> GSM720352</td>
<td> MCF-7</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683918</td>
<td> SK-N-MC</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683898</td>
<td> HCT-116</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683827</td>
<td> NB4</td>
<td>cancer
</td></tr>
<tr>
<td> GSM720351</td>
<td> MCF-7</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683795</td>
<td> HepG2</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683793</td>
<td> MCF-7</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683855</td>
<td> HepG2</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683924</td>
<td> LNCaP</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683909</td>
<td> PFSK-1</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683866</td>
<td> PANC-1</td>
<td>cancer, male
</td></tr>
<tr>
<td> GSM683800</td>
<td> SK-N-SH_RA</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683920</td>
<td> PANC-1</td>
<td>cancer, male
</td></tr>
<tr>
<td> GSM720355</td>
<td> HEK293</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683926</td>
<td> HeLa-S3</td>
<td>cancer
</td></tr>
<tr>
<td> GSM980576</td>
<td> HL-60</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683768</td>
<td> LNCaP</td>
<td>cancer
</td></tr>
<tr>
<td> GSM980574</td>
<td> MCF-7</td>
<td>cancer
</td></tr>
<tr>
<td> GSM720349</td>
<td> NT2-D1</td>
<td>cancer
</td></tr>
<tr>
<td> GSM1048457</td>
<td> HepG2</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683787</td>
<td> MCF-7</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683825</td>
<td> HEK293</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683862</td>
<td> LNCaP</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683832</td>
<td> NT2-D1</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683776</td>
<td> LNCaP</td>
<td>cancer
</td></tr>
<tr>
<td> GSM980581</td>
<td> UCH-1</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683929</td>
<td> NT2-D1</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683804</td>
<td> NT2-D1</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683780</td>
<td> K562</td>
<td>cancer
</td></tr>
<tr>
<td> GSM720350</td>
<td> MCF-7</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683919</td>
<td> SK-N-SH_RA</td>
<td>cancer
</td></tr>
<tr>
<td> GSM683810</td>
<td> PFSK-1</td>
<td>cancer
</td></tr></tbody></table>
<h3> <span class="mw-headline" id="Generate_methylation_matrices_over_Informative_Bins">Generate methylation matrices over Informative Bins </span></h3>
<ul><li> Majority of informative bins are unmethylated in normal tissues
</li><li> Two particular cell lines were more methylated: 
<ul><li>HTR: Trophoblast (HTR-8/SVneo) cell line. A thin layer of ectoderm that forms the wall of many mammalian blastulas and functions in the nutrition and implantation of the embryo, cell karyotype: normal, cell lineage: Blastula
</li><li>GM12878-XiMat: lymphoblastoid, cell karyotype: relatively normal, cell lineage: International HapMap Project - CEPH/Utah - European Caucasion; Epstein-Barr Virus;Cloned for maternal X inactivation
</li></ul>
</li><li> Heatmap (normal tissues, showing replicates)
</li></ul>
<pre>Colon cancer
<a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:ColonCancerBins_normalTissues_heatmap.png" class="image"><img alt="ColonCancerBins normalTissues heatmap.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/ColonCancerBins_normalTissues_heatmap.png" width="600" height="300"></a>
Lung cancer
<a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:LungCancerBins_normalTissues_heatmap.png" class="image"><img alt="LungCancerBins normalTissues heatmap.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/LungCancerBins_normalTissues_heatmap.png" width="600" height="300"></a>
Pancreatic cancer
<a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:PancreaticCancerBins_normalTissues_heatmap.png" class="image"><img alt="PancreaticCancerBins normalTissues heatmap.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/PancreaticCancerBins_normalTissues_heatmap.png" width="600" height="300"></a>
</pre>
<ul><li> Heatmaps ( all replicates were merged into one file )
</li></ul>
<pre>Colon cancer: normal tissues, cancer cell lines
<a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:ColonCancerBins_normalTissues_heatmap_mergedReps.png" class="image"><img alt="ColonCancerBins normalTissues heatmap mergedReps.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/ColonCancerBins_normalTissues_heatmap_mergedReps.png" width="600" height="419"></a><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:ColonCancerBins_cancerCellLines_heatmap_mergedReps.png" class="image"><img alt="ColonCancerBins cancerCellLines heatmap mergedReps.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/ColonCancerBins_cancerCellLines_heatmap_mergedReps.png" width="600" height="419"></a>
Lung cancer: normal tissues, cancer cell lines
<a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:LungCancerBins_normalTissues_heatmap_mergedReps.png" class="image"><img alt="LungCancerBins normalTissues heatmap mergedReps.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/LungCancerBins_normalTissues_heatmap_mergedReps.png" width="600" height="419"></a><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:LungCancerBins_cancerCellLines_heatmap_mergedReps.png" class="image"><img alt="LungCancerBins cancerCellLines heatmap mergedReps.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/LungCancerBins_cancerCellLines_heatmap_mergedReps.png" width="600" height="419"></a>
Pancreatic cancer: normal tissues, cancer cell lines
<a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:PancreaticCancerBins_normalTissues_heatmap_mergedReps.png" class="image"><img alt="PancreaticCancerBins normalTissues heatmap mergedReps.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/PancreaticCancerBins_normalTissues_heatmap_mergedReps.png" width="600" height="419"></a><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:PancreaticCancerBins_cancerCellLines_heatmap_mergedReps.png" class="image"><img alt="PancreaticCancerBins cancerCellLines heatmap mergedReps.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/PancreaticCancerBins_cancerCellLines_heatmap_mergedReps.png" width="600" height="419"></a>
</pre>
<h3> <span class="mw-headline" id=".22Normal.22_reference_bins">"Normal" reference bins</span></h3>
<ul><li> Dr. Zhang suggested to use only "BC_" samples that we are sure are 100% normal tissues. There are 17 of these. After plotting the heatmaps, placenta and testis should be removed. 
</li><li> Median methylation levels across normal tissues (removed placenta, testis)
<ul><li> <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Colon.bins.median-methylation-normal.txt" title="File:Colon.bins.median-methylation-normal.txt">File:Colon.bins.median-methylation-normal.txt</a> <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Colon.heatmap.qqplot.pdf" title="File:Colon.heatmap.qqplot.pdf">File:Colon.heatmap.qqplot.pdf</a>
</li><li> <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Lung.bins.median-methylation-normal.txt" title="File:Lung.bins.median-methylation-normal.txt">File:Lung.bins.median-methylation-normal.txt</a> <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Lung.heatmap.qqplot.pdf" title="File:Lung.heatmap.qqplot.pdf">File:Lung.heatmap.qqplot.pdf</a>
</li><li> <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Pancreatic.bins.median-methylation-normal.txt" title="File:Pancreatic.bins.median-methylation-normal.txt">File:Pancreatic.bins.median-methylation-normal.txt</a> <a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Pancreatic.heatmap.qqplot.pdf" title="File:Pancreatic.heatmap.qqplot.pdf">File:Pancreatic.heatmap.qqplot.pdf</a>
</li></ul>
</li></ul>
<p><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/File:Median-methylation-at-informative-bins-in-normal.png" class="image"><img alt="Median-methylation-at-informative-bins-in-normal.png" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/Median-methylation-at-informative-bins-in-normal.png" width="300" height="569"></a>
</p>
<ul><li> To get the "normal" reference:
</li></ul>
<table class="wikitable">
<tbody><tr>
<td align="center" style="background:#f0f0f0;">'
</td>
<td align="center" style="background:#f0f0f0;"><b>Colon</b>
</td>
<td align="center" style="background:#f0f0f0;"><b>Lung</b>
</td>
<td align="center" style="background:#f0f0f0;"><b>Pancreatic</b>
</td></tr>
<tr>
<td> Number of starting bins</td>
<td>2274</td>
<td>1144</td>
<td>1699
</td></tr>
<tr>
<td> Number of bins with &gt;=15 normal tissues coverage</td>
<td>2138</td>
<td>1068</td>
<td>1622
</td></tr>
<tr>
<td> Number of bins with median methylation in normal &lt;0.2</td>
<td>2070 (96.82%)</td>
<td>1017(95.22%)</td>
<td>1597(98.46%)
</td></tr>
<tr>
<td> Number of bins with median methylation in normal &lt;0.1</td>
<td>1944(90.93%)</td>
<td>932(87.27%)</td>
<td>1504(92.73%)
</td></tr>
<tr>
<td> Number of bins with median methylation in normal &lt;0.05</td>
<td>1677(78.43%)</td>
<td>716(67.04%)</td>
<td>1238(76.33%)
</td></tr>
<tr>
<td>&nbsp;%bins(out of all orig bins, if 0.2 cutoff)</td>
<td>91.03%</td>
<td>88.90%</td>
<td>94.00%
</td></tr></tbody></table>
<h2> <span class="mw-headline" id="Divide_haplotypes_into_clusters_in_solid_tumor_samples"> Divide haplotypes into clusters in solid tumor samples </span></h2>
<ul><li> Go back to solid tumor data and try to identify haplotype clusters. 
</li><li> Try to estimate tumor purify based on the fraction of cancer haplotypes (these should be partially or fully methylated).
</li></ul>
<h2> <span class="mw-headline" id="Discussion"> Discussion </span></h2>
<ul><li> Discussion with Dr. Zhang (9/12/14): 
<ul><li> Intersect informative bins with blood UMRs, what are those regions enriched for in GREAT? (DONE)
</li><li> How to tell whether the minor or major component is the tumor DNA&nbsp;? It appears that these tumor references have varying amount of minor component. 
<ul><li> Ave. methyl colon cancer tumors: major = 0.7348, minor = 0.1602
</li><li> Ave. methyl lung cancer tumors:  major = 0.6393, minor = 0.2574
</li><li> Ave. methyl pancreatic cancer tumors: major = 0.7623, minor = 0.1181
</li><li> Although majority of informative CGIs are overlapping with blood UMRs, the minor components are not always unmethylated, thus there may be other normal cells in the mixture.
</li></ul>
</li><li> Use MethylPurify to generate cancer reference from primary tumors. So we can create a database of true cancer references (after substracting out normal cells). We can use the pure cancer reference for methylation haplotyping to detect cancer reads in blood.
</li></ul>
</li></ul>

<!-- 
NewPP limit report
Preprocessor node count: 42/1000000
Post-expand include size: 0/2097152 bytes
Template argument size: 0/2097152 bytes
Expensive parser function count: 0/100
-->

<!-- Saved in parser cache with key wikidb-LabNotes:pcache:idhash:19089-0!*!*!!en!2!* and timestamp 20151020074840 -->
</div><div class="printfooter">
Retrieved from "<a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&oldid=62846">http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&amp;oldid=62846</a>"</div>
		<div id="catlinks" class="catlinks catlinks-allhidden"></div>		<!-- end content -->
				<div class="visualClear"></div>
	</div>
</div></div>
<div id="column-one">
	<div id="p-cactions" class="portlet">
		<h5>Views</h5>
		<div class="pBody">
			<ul>
				<li id="ca-nstab-main" class="selected"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Dinh/Dinh_2014/NOTES/2014-9-15" title="View the content page [alt-c]" accesskey="c">Page</a></li>
				<li id="ca-talk" class="new"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Talk:Dinh/Dinh_2014/NOTES/2014-9-15&action=edit&redlink=1" title="Discussion about the content page [alt-t]" accesskey="t">Discussion</a></li>
				<li id="ca-edit"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&action=edit" title="You can edit this page. Please use the preview button before saving [alt-e]" accesskey="e">Edit</a></li>
				<li id="ca-history"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&action=history" title="Past revisions of this page [alt-h]" accesskey="h">History</a></li>
				<li id="ca-move"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Special:MovePage/Dinh/Dinh_2014/NOTES/2014-9-15" title="Move this page [alt-m]" accesskey="m">Move</a></li>
				<li id="ca-watch"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&action=watch&token=99119ade0107ac695d70b2cdc631935b%2B%5C" title="Add this page to your watchlist [alt-w]" accesskey="w">Watch</a></li>
			</ul>
		</div>
	</div>
	<div class="portlet" id="p-personal">
		<h5>Personal tools</h5>
		<div class="pBody">
			<ul>
				<li id="pt-userpage"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/User:Shicheng" class="new" title="Your user page [alt-.]" accesskey=".">Shicheng</a></li>
				<li id="pt-mytalk"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/User_talk:Shicheng" class="new" title="Your talk page [alt-n]" accesskey="n">My talk</a></li>
				<li id="pt-preferences"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Special:Preferences" title="Your preferences">My preferences</a></li>
				<li id="pt-watchlist"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Special:Watchlist" title="A list of pages you are monitoring for changes [alt-l]" accesskey="l">My watchlist</a></li>
				<li id="pt-mycontris"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Special:Contributions/Shicheng" title="A list of your contributions [alt-y]" accesskey="y">My contributions</a></li>
				<li id="pt-logout"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Special:UserLogout&returnto=Dinh%2FDinh+2014%2FNOTES%2F2014-9-15&returntoquery=printable%3Dyes" title="Log out">Log out</a></li>
			</ul>
		</div>
	</div>
	<div class="portlet" id="p-logo">
		<a title="Visit the main page" style="background-image: url(/common/images/umd.gif);" href="http://genome-tech.ucsd.edu/LabNotes/index.php/Main_Page"></a>
	</div>
	<script type="text/javascript"> if (window.isMSIE55) fixalpha(); </script>
	<div class="generated-sidebar portlet" id="p-navigation">
		<h5>Navigation</h5>
		<div class="pBody">
			<ul>
				<li id="n-mainpage"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Main_Page" title="Visit the main page [alt-z]" accesskey="z">Main Page</a></li>
				<li id="n-currentevents"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/ZhangLabWiki:Current_events" title="Find background information on current events">Current events</a></li>
				<li id="n-recentchanges"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Special:RecentChanges" title="A list of recent changes in the wiki [alt-r]" accesskey="r">Recent changes</a></li>
				<li id="n-randompage"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Special:Random" title="Load a random page [alt-x]" accesskey="x">Random page</a></li>
			</ul>
		</div>
	</div>
	<div class="generated-sidebar portlet" id="p-Investigators">
		<h5>Investigators</h5>
		<div class="pBody">
			<ul>
				<li id="n-Matt-Cai"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Matthew_Cai">Matt Cai</a></li>
				<li id="n-Song-Chen"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Song_Chen">Song Chen</a></li>
				<li id="n-Eric-Chu"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Eric_Chu">Eric Chu</a></li>
				<li id="n-Dinh-Diep"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Dinh_Diep">Dinh Diep</a></li>
				<li id="n-Shicheng-Guo"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Shicheng_Guo">Shicheng Guo</a></li>
				<li id="n-Alan-Fung"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Alan_Fung">Alan Fung</a></li>
				<li id="n-Daniel-Jacobsen"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Daniel_Jacobsen">Daniel Jacobsen</a></li>
				<li id="n-Blue-Lake"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Blue_Lake">Blue Lake</a></li>
				<li id="n-Ho-Suk-Lee"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Ho_Suk_Lee">Ho Suk Lee</a></li>
				<li id="n-Alice-Li"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Alice_Li">Alice Li</a></li>
				<li id="n-Andrew-Richards"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Andrew_Richards">Andrew Richards</a></li>
				<li id="n-Brandon-Sos"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Brandon_Sos">Brandon Sos</a></li>
				<li id="n-Chris-Wei"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Chris_Wei">Chris Wei</a></li>
				<li id="n-Kun-Zhang"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Kun_Zhang">Kun Zhang</a></li>
			</ul>
		</div>
	</div>
	<div id="p-search" class="portlet">
		<h5><label for="searchInput">Search</label></h5>
		<div id="searchBody" class="pBody">
			<form action="http://genome-tech.ucsd.edu/LabNotes/index.php" id="searchform">
				<input type="hidden" name="title" value="Special:Search">
				<input type="search" name="search" title="Search ZhangLabWiki [alt-f]" accesskey="f" id="searchInput">
				<input type="submit" name="go" value="Go" title="Go to a page with this exact name if exists" id="searchGoButton" class="searchButton">&nbsp;
				<input type="submit" name="fulltext" value="Search" title="Search the pages for this text" id="mw-searchButton" class="searchButton">
			</form>
		</div>
	</div>
	<div class="portlet" id="p-tb">
		<h5>Toolbox</h5>
		<div class="pBody">
			<ul>
				<li id="t-whatlinkshere"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Special:WhatLinksHere/Dinh/Dinh_2014/NOTES/2014-9-15" title="A list of all wiki pages that link here [alt-j]" accesskey="j">What links here</a></li>
				<li id="t-recentchangeslinked"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Special:RecentChangesLinked/Dinh/Dinh_2014/NOTES/2014-9-15" title="Recent changes in pages linked from this page [alt-k]" accesskey="k">Related changes</a></li>
				<li id="t-upload"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Special:Upload" title="Upload files [alt-u]" accesskey="u">Upload file</a></li>
				<li id="t-specialpages"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/Special:SpecialPages" title="A list of all special pages [alt-q]" accesskey="q">Special pages</a></li>
				<li id="t-permalink"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php?title=Dinh/Dinh_2014/NOTES/2014-9-15&oldid=62846" title="Permanent link to this revision of the page">Permanent link</a></li>
			</ul>
		</div>
	</div>
</div><!-- end of the left (by default at least) column -->
<div class="visualClear"></div>
<div id="footer">
	<div id="f-copyrightico">
		<a href="http://www.gnu.org/copyleft/fdl.html"><img src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/gnu-fdl.png" alt="GNU Free Documentation License 1.2" width="88" height="31"></a>
	</div>
	<div id="f-poweredbyico">
		<a href="http://www.mediawiki.org/"><img src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/poweredby_mediawiki_88x31.png" alt="Powered by MediaWiki" width="88" height="31"></a>
	</div>
	<ul id="f-list">
		<li id="lastmod"> This page was last modified on 7 October 2014, at 19:13.</li>
		<li id="viewcount">This page has been accessed 131 times.</li>
		<li id="copyright">Content is available under <a class="external" href="http://www.gnu.org/copyleft/fdl.html">GNU Free Documentation License 1.2</a>.</li>
		<li id="privacy"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/ZhangLabWiki:Privacy_policy" title="ZhangLabWiki:Privacy policy">Privacy policy</a></li>
		<li id="about"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/ZhangLabWiki:About" title="ZhangLabWiki:About">About ZhangLabWiki</a></li>
		<li id="disclaimer"><a href="http://genome-tech.ucsd.edu/LabNotes/index.php/ZhangLabWiki:General_disclaimer" title="ZhangLabWiki:General disclaimer">Disclaimers</a></li>
	</ul>
</div>
</div>
<script>if(window.mw){
	mw.loader.load(["mediawiki.user", "mediawiki.util", "mediawiki.page.ready", "mediawiki.legacy.wikibits", "mediawiki.legacy.ajax", "mediawiki.action.watch.ajax"]);
}
</script><script type="text/javascript" src="./Dinh_Dinh 2014_NOTES_2014-9-15 - ZhangLabWiki_files/load(4).php"></script>
<script>if(window.mw){
	mw.user.options.set({"ccmeonemails":0,"cols":80,"date":"default","diffonly":0,"disablemail":0,"disablesuggest":0,"editfont":"default","editondblclick":0,"editsection":1,"editsectiononrightclick":0,"enotifminoredits":0,"enotifrevealaddr":0,"enotifusertalkpages":1,"enotifwatchlistpages":0,"extendwatchlist":0,"externaldiff":0,"externaleditor":0,"fancysig":0,"forceeditsummary":0,"gender":"male","hideminor":0,"hidepatrolled":0,"highlightbroken":1,"imagesize":2,"justify":0,"math":1,"minordefault":0,"newpageshidepatrolled":0,"nocache":0,"noconvertlink":0,"norollbackdiff":0,"numberheadings":0,"previewonfirst":0,"previewontop":1,"quickbar":5,"rcdays":7,"rclimit":50,"rememberpassword":"1","rows":25,"searchlimit":20,"showhiddencats":0,"showjumplinks":1,"shownumberswatching":1,"showtoc":1,"showtoolbar":1,"skin":"monobook","stubthreshold":0,"thumbsize":2,"underline":2,"uselivepreview":0,"usenewrc":0,"watchcreations":0,"watchdefault":0,"watchdeletion":0,"watchlistdays":3,"watchlisthideanons":0,
	"watchlisthidebots":0,"watchlisthideliu":0,"watchlisthideminor":0,"watchlisthideown":0,"watchlisthidepatrolled":0,"watchmoves":0,"wllimit":250,"variant":"en","language":"en","searchNs0":true,"searchNs1":false,"searchNs2":false,"searchNs3":false,"searchNs4":false,"searchNs5":false,"searchNs6":false,"searchNs7":false,"searchNs8":false,"searchNs9":false,"searchNs10":false,"searchNs11":false,"searchNs12":false,"searchNs13":false,"searchNs14":false,"searchNs15":false,"searchNs100":false,"searchNs101":false,"searchNs102":false,"searchNs103":false,"searchNs104":false,"searchNs105":false,"searchNs107":false,"searchNs108":false,"searchNs109":false,"searchNs110":false,"searchNs111":false,"searchNs112":false,"searchNs113":false,"searchNs114":false,"searchNs115":false,"searchNs116":false,"searchNs117":false,"searchNs118":false,"searchNs119":false,"searchNs120":false,"searchNs121":false,"searchNs122":false,"searchNs123":false,"searchNs124":false,"searchNs125":false,"searchNs126":false,"nickname":""
	,"timecorrection":"Offset|0","watchlisttoken":"4d66fd5215e401c3556112983f334af8cb72c70c"});;mw.user.tokens.set({"editToken":"5dee11b0b2fdbad14cdcfc144fb8b4cb+\\","watchToken":"599567381168be17a726310498d85054+\\"});;mw.loader.state({"user.options":"ready","user.tokens":"ready"});
	
	/* cache key: wikidb-LabNotes:resourceloader:filter:minify-js:4:a38cd885d8ec5e9dc3d8e7795af7d9b6 */
}
</script><!-- Served in 0.168 secs. --></body></html>