#!/bin/bash
#PBS -N SRR10070127
#PBS -q shortq
#PBS -l nodes=1:ppn=6
#PBS -M Guo.Shicheng@marshfieldresearch.org
#PBS -m abe
cd /mnt/bigdata/Genetic/Projects/shg047/methylation/pancrease/WGBS
# fastq-dump --skip-technical --split-files --gzip SRR10070127
trim_galore --paired --phred33 --clip_R1 3 --clip_R2 6 --fastqc --illumina SRR10070127_1.fastq.gz SRR10070127_2.fastq.gz --output_dir ../fastq_trim
bismark --bowtie2 --multicore 2 --fastq -N 1 --phred33-quals ~/hpc/db/hg19/bismark -1 ../fastq_trim/SRR10070127_1_val_1.fq.gz -2 ../fastq_trim/SRR10070127_2_val_2.fq.gz -o ../bam
filter_non_conversion --paired ../bam/SRR10070127_1_val_1_bismark_bt2_pe.bam
deduplicate_bismark --bam ../bam/SRR10070127_1_val_1_bismark_bt2.nonCG_filtered.bam
samtools sort -@ 8 ../bam/SRR10070127_1_val_1_bismark_bt2.nonCG_filtered.deduplicated.bam -o ../sortbam/SRR10070127_bismark_bt2_pe.sortc.bam
samtools index ../sortbam/SRR10070127_bismark_bt2_pe.sortc.bam
cd ../sortbam
perl ~/bin/samInfoPrep4Bam2Hapinfo.pl ~/oasis/db/hg19/hg19.cut10k.bed > saminfo.txt
perl ~/bin/bam2hapInfo2PBS.pl saminfo.txt submit bismark ~/hpc/db/hg19/hg19.chrom.sizes ~/hpc/db/hg19/HsGenome19.CpG.positions.txt
bismark_methylation_extractor --no_overlap --multicore 2 --merge_non_CpG --bedGraph --cutoff 5 --ignore 1 --buffer_size 4G --comprehensive --output ../methyfreq  ../bam/SRR10070127_1_val_1_bismark_bt2.nonCG_filtered.deduplicated.bam
coverage2cytosine --merge_CpG --gzip --genome_folder ~/hpc/db/hg19/bismark -o SRR10070127_1.mergeCpG.bed SRR10070127_1_val_1_bismark_bt2_pe.nonCG_filtered.bismark.cov.gz
