---
title: "Heatmap Analysis to Genome-wide MHL Matrix in MONOD Project"
author: "Shicheng Guo"
date: "Friday, December 18, 2015"
output: html_document
---
  
  Update: Jan-19-2015

```{r}
library("impute")
RawNARemove<-function(data,missratio=0.3){
  threshold<-(missratio)*dim(data)[2]
  NaRaw<-which(apply(data,1,function(x) sum(is.na(x))>threshold))
  zero<-which(apply(data,1,function(x) all(x==0))==T)
  NaRAW<-c(NaRaw,zero)
  if(length(NaRAW)>0){
    dat<-data[-NaRAW,]
  }else{
    dat<-data;
  }
  dat
} 
ci95<-function(x){
  error <- qt(0.975,df=length(x)-1)*sd(x)/sqrt(length(x))
  m<-round(mean(x),2)
  d<-round(mean(x)-error,2)
  u<-round(mean(x)+error,2)
  paste(m, "(",d,"-",u,")",sep="")
}
ttestPValue<-function(data,x1,x2){
  data<-data+matrix(rnorm(nrow(data)*ncol(data),0,0.00000000001),nrow=nrow(data),ncol=ncol(data))
  output<-matrix(NA,dim(data)[1],5)
  for(i in 1:dim(data)[1]){
    out<-data.frame()
    if(all(! any(all(is.na(data[i,x1])),all(is.na(data[i,x2]))),sum(is.na(data[i,]))<0.5*length(data[i,]))){ 
      tmp1<-try(t.test(as.numeric(data[i,x1]),as.numeric(data[i,x2]),paired=F, na.action=na.omit))
      output[i,1]<-tmp1$p.value
      output [i,2]<-mean(as.numeric(data[i,x1]))-mean(as.numeric(data[i,x2]))
      output[i,3]<-"t-test"
      output[i,4]<-mean(as.numeric(data[i,x1]))
      output[i,5]<-mean(as.numeric(data[i,x2]))
    }
  }
  FDR<-p.adjust(output[,1],"fdr")
  out<-data.frame(rownames(data),output,FDR)
  colnames(out)<-c("ID","P-value","Delta(M1-M2)","Test","Mean1","Mean2","FDR")
  out
}
ColNARemove<-function(data,missratio=0.3){
  threshold<-(missratio)*dim(data)[1]
  NaCol<-which(apply(data,2,function(x) sum(is.na(x))>threshold))
  zero<-which(apply(data,2,function(x) all(x==0))==T)
  NaCOL<-c(NaCol,zero)
  if(length(NaCOL)>0){
    data1<-data[,-NaCOL]
  }else{
    data1<-data;
  }
  data1
}
bedwithgap<-function(bed,gap){
  bed<-as.matrix(bed)
  bed[,2]=as.numeric(bed[,2])-gap
  bed[,3]=as.numeric(bed[,3])+gap
  bed<-data.frame(bed)
  bed
}
Rbedtools<-function(functionstring="intersectBed",bed1,bed2,opt.string=""){
  #create temp files
  a.file=tempfile()
  b.file=tempfile()
  out   =tempfile()
  options(scipen =99) # not to use scientific notation when writing out
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  write.table(bed2,file=b.file,quote=F,sep="\t",col.names=F,row.names=F)
  # create the command string and call the command using system()
  command=paste(functionstring,"-a",a.file,"-b",b.file,opt.string,">",out,sep=" ")
  cat(command,"\n")
  try(system(command))
  res=read.table(out,header=F)
  unlink(a.file);unlink(b.file);unlink(out)
  return(res)
}
cor2bed<-function(cor){
  a<-unlist(lapply(strsplit(as.character(cor),split=c(":")),function(x) strsplit(x,"-")))
  bed<-matrix(a,ncol=3,byrow=T)
  return(data.frame(bed))
}
bed2cor<-function(bed){
  cor<-apply(bed,1,function(x){paste(unlist(strsplit(x,"\t"))[1],":",unlist(strsplit(x,"\t"))[2],"-",unlist(strsplit(x,"\t"))[3],sep="")})
  cor<-gsub("[ ]","",cor)
  return(cor)
}
PCAPlot<-function(data,pheno,output,multifigure=T){
  pca <- prcomp(data,center=T,scale = F)  # Here, input file: row is individual and column is variable
  outputfile=paste(output,".pdf",sep="")
  pdf(outputfile)
  if(multifigure){
    par(mfrow=c(2,2),mar=c(4,4,4,4))  
  }
  plot((pca$sdev[1:10])^2,type="o",xaxt="n",ylab="Variances",xlab="Principle Components",col="red",lwd=2)
  axis(1,at=0:10,labels=paste("PC",0:10,sep=""))
  var<-c()
  for(i in 1:length(pca$sdev)){var[i]<-sum((pca$sdev[1:i])^2)/sum((pca$sdev)^2)}
  plot(var,ylab="total variance",xlab="number of principle components",lwd=2,type="l")
  abline(h=0.8,col="grey",lty=2)
  abline(v=which(var>0.8)[1],col="grey",lty=2)
  scores <- data.frame(pheno, pca$x[,1:3])
  col = as.numeric(as.factor(pheno))
  plot(x=scores$PC1,y=scores$PC2, xlim=c(min(scores$PC1),max(scores$PC1)),ylim=c(min(scores$PC2),max(scores$PC2)),type="n",xlab="PC1",ylab="PC2")
  for(i in 1:length(scores$PC1)){
    points(scores$PC1[i],scores$PC2[i],pch=as.numeric(as.factor(pheno))[i],col=col[i],cex=0.8,lwd=2)
  }
  legend("bottomright",legend=names(table(pheno)),pch=1:length(table(pheno)),col=1:length(table(pheno)),bty="n")
  plot(x=scores$PC1,y=scores$PC3, xlim=c(min(scores$PC1),max(scores$PC1)),ylim=c(min(scores$PC3),max(scores$PC3)),type="n",xlab="PC1",ylab="PC3")
  for(i in 1:length(scores$PC1)){
    points(scores$PC1[i],scores$PC3[i],pch=as.numeric(as.factor(pheno))[i],col=col[i],cex=0.9,lwd=2)
  }
  legend("bottomright",legend=names(table(pheno)),pch=1:length(table(pheno)),col=1:length(table(pheno)),bty="n")
  dev.off()
}
gsi<-function(data){
  group=names(table(colnames(data)))
  index=colnames(data)
  gsi<-c()
  gmaxgroup<-c()
  for(i in 1:nrow(data)){
    gsit<-0
    gmax<-names(which.max(tapply(as.numeric(data[i,]),index,mean)))
    for(j in 1:length(group)){
      tmp<-(1-10^(mean(data[i,][which(index==group[j])]))/10^(mean(data[i,][which(index==gmax)])))/(length(group)-1)
      gsit<-gsit+tmp
    }
    gmaxgroup<-c(gmaxgroup,gmax)
    gsi<-c(gsi,gsit)
    print(c(gmax,gsit))
  }
  rlt=data.frame(region=rownames(data),group=gmaxgroup,GSI=gsi)
  return(rlt)
}
```


```{r}
setwd("/home/shg047/monod/dec")
infile="WGBS_methHap_load_matrix_20Oct2015.txt"
file1<-read.table(infile,head=T,sep="\t",row.names=1,as.is=T,check.names=F)

# miss value detection and imputation
library("impute")
scol<-grep("N37-|STL|HCT|methylC|tumor_lung|tumor_colon|normal_colon|normal_lung|adenocarcinoma_lung",colnames(data))
newdata<-data[,scol]
newdata<-RawNARemove(newdata,missratio=0.3)
nrow(newdata)/nrow(data)

newdata<-scale(newdata)
d <- dist(t(newdata), method = "euclidean") # distance matrix
pdf("hist.pdf")
fit <- hclust(d, method="ward") 
plot(fit,cex=0.5) # display dendogram
dev.off()

f2<-impute.knn(data.matrix(f2))$data

library("preprocessCore")
f2.t1<-normalize.quantiles(f2[,13:58])
library("sva")
batch=c(rep(1,10),rep(2,36))
f2.t2<-ComBat(f2.t1, batch, mod=NULL, par.prior = TRUE,prior.plots = FALSE)
f2[,13:58]<-f2.t2

# re-plot the cluster by sequencing sample Tag
colnames(f2)<-colnames(file1)
f3 <- t(na.omit(f2)) # listwise deletion of missing
histFfile=paste("Figure.Cluster.mhl.Dendrogram.combat.seqTag.ward.D.plot",infile,"pdf",sep=".")
d <- dist(f3, method = "euclidean") # distance matrix
fit <- hclust(d, method="ward.D") 
pdf(histFfile)
plot(fit,cex=0.5,hang=-1) # display dendogram
dev.off()
histFfile=paste("Figure.Cluster.mhl.Dendrogram.combat.seqTag.plot.average",infile,"pdf",sep=".")
d <- dist(f3, method = "euclidean") # distance matrix
fit <- hclust(d, method="average") 
pdf(histFfile)
plot(fit,cex=0.5,hang=-1) # display dendogram
dev.off()

# re-assign colnames
colnames(f2) 
colnames(f2)<-gsub("_","-",colnames(f2))
colname2<-unlist(lapply(colnames(f2),function(x) unlist(strsplit(x,"[.]"))[1]))
colname2
colnames(f2)<-colname2
# be sure all the sample information has been stored in the following database
saminfo2<-read.table("/home/shg047/monod/phase2/newsaminfo.txt",head=T,sep="\t",as.is=T)
saminfo2<-saminfo2[match(colname2,saminfo2[,1]),]
saminfo2
colnames(f2)<-saminfo2[,2]
colnames(f2)

# check the dendrogram after missing, quantile
f3 <- t(na.omit(f2)) # listwise deletion of missing
# f3 <- scale(f3) # standardize variables
sum(is.na(file1))/(nrow(file1)*ncol(file1))
sum(is.na(f2))/(nrow(f2)*ncol(f2))
histFfile=paste("Figure.hist.combat",infile,"pdf",sep=".")
pdf(histFfile)
boxplot(f2,outline=F,horizontal=T,notch=F,las=1,cex.axis=0.65,col="blue")
dev.off()
histFfile=paste("Figure.Cluster.mhl.Dendrogram.combat.ward.D.plot",infile,"pdf",sep=".")
d <- dist(f3, method = "euclidean") # distance matrix
fit <- hclust(d, method="ward.D") 
pdf(histFfile)
plot(fit,cex=0.5,hang=-1) # display dendogram
dev.off()
histFfile=paste("Figure.Cluster.mhl.Dendrogram.combat.plot.average",infile,"pdf",sep=".")
d <- dist(f3, method = "euclidean") # distance matrix
fit <- hclust(d, method="average") 
pdf(histFfile)
plot(fit,cex=0.5,hang=-1) # display dendogram
dev.off()


# tissue specific hml, 1360 regions
gsi_1<-gsi(f2[,13:58])
gsi_2<-gsi(f2)
data<-subset(gsi_1,gsi_1[,3]>0.5)
table(data[,2])
z<-c()
for (i in names(table(data[2]))){
  x<-subset(data,data[,2]==i)
  y<-x[order(x[,3],decreasing=T)[1:80],]
  z<-rbind(z,y)
}
z<-na.omit(z)
dim(z)
write.table(z,file="supplementary.1360.tsi.txt",col.names=T,row.names=F,sep="\t",quote=F)

gsi1<-subset(gsi_1,GSI>0.6)
gsi2<-subset(gsi_2,GSI>0.6)

dim(gsi1)
dim(gsi2)

write.table(gsi1,file="Table.GSI.WGBS.Remove.H1.WBC.rlt.txt",col.names=T,row.names=F,quote=F,sep="\t")
write.table(gsi2,file="Table.GSI.WGBS.with.H1.WBC.rlt.txt",col.names=T,row.names=F,quote=F,sep="\t")
```

```{r,echo=TRUE}
pdf("unsupervised.cluster.analysis.based.on.wgbs.mhl.txt")
plot(hclust(dist(t(f2))))
dev.off()
```


Then we can plot the heatmap based on high GSI regions to the total 41 tissue samples (without H1, Cancer and WBC)
```{r}
library("gplots")
library("RColorBrewer")
library("grDevices")
mydata1<-f2[which(gsi_1[,3]>0.6),]
hclustfunc <- function(x) hclust(x, method="ward.D")
distfunc <- function(x) dist(x, method="euclidean")
# perform clustering on rows and columns
cl.row <- hclustfunc(distfunc(mydata1))
cl.col <- hclustfunc(distfunc(t(mydata1)))
# extract cluster assignments; i.e. k=8 (rows) k=5 (columns)
gr.row <- cutree(cl.row, 6)
# gr.col <- cutree(cl.col, 5)
# require(RColorBrewer)
col1 <- brewer.pal(6, "Set1")     # the maximum of brewer.pal is 12
#col2 <- brewer.pal(5, "Pastel1")
tol21rainbow= c("#771155", "#AA4488", "#CC99BB", "#114477", "#4477AA", "#77AADD", "#117777", "#44AAAA", "#77CCCC", "#117744", "#44AA77", "#88CCAA", "#777711", "#AAAA44", "#DDDD77", "#774411", "#AA7744", "#DDAA77", "#771122", "#AA4455", "#DD7788")
col2<-tol21rainbow[as.numeric(as.factor(cl.col$labels))]

library("grDevices")
library("gplots")
mydata1[mydata1<0]<-0
mydata1[mydata1>1]<-1
col=colorRampPalette(c("blue", "yellow"))(20) 

pdf("heatmap.wgbs.mhl.gsi0.6.without.H1.wbc.cancer.pdf")
heatmap.2(as.matrix(mydata1),hclustfun=hclustfunc, distfun=distfunc,
          RowSideColors=col1[gr.row], 
          ColSideColors=col2,
          Colv=T,Rowv=T,key=T,
          col=col,
          cexCol=0.65,
          keysize=1,
          labRow=NA,
          trace="none",
          density.info="none")
dev.off()
```

Then we can plot the heatmap based on high GSI regions to the total 61 (with H1, Cancer and WBC)

```{r}
library("gplots")
library("RColorBrewer")
library("grDevices")
mydata1<-f2[which(gsi_2[,3]>0.6),]
mydata<-head(mydata1)
hclustfunc <- function(x) hclust(x, method="ward.D")
distfunc <- function(x) dist(x, method="euclidean")
# perform clustering on rows and columns
cl.row <- hclustfunc(distfunc(mydata))
cl.col <- hclustfunc(distfunc(t(mydata)))
# extract cluster assignments; i.e. k=8 (rows) k=5 (columns), ncut should smaller than total color number. "Set1" of brewer.pal <12
ncut=6  
# gr.col <- cutree(cl.col, 5)
require(RColorBrewer)
col1 <- brewer.pal(ncut, "Set1")[gr.row[cl.row$order]]     # the maximum of brewer.pal is 12
#col2 <- brewer.pal(5, "Pastel1")
tol21rainbow= c("#771155", "#AA4488", "#CC99BB", "#114477", "#4477AA", "#77AADD", "#117777", "#44AAAA", "#77CCCC", "#117744", "#44AA77", "#88CCAA", "#777711", "#AAAA44", "#DDDD77", "#774411", "#AA7744", "#DDAA77", "#771122", "#AA4455", "#DD7788")
coll<-gr.row[as.numeric(as.factor(cl.row$labels))]

col2<-tol21rainbow[as.numeric(as.factor(cl.col$labels))]
library("grDevices")
library("gplots")
mydata[mydata<0]<-0
mydata[mydata>1]<-1
col=colorRampPalette(c("blue", "yellow"))(20) 
pdf("heatmap.wgbs.mhl.gsi0.6.with.H1.wbc.cancer.pdf")
heatmap.2(as.matrix(mydata),hclustfun=hclustfunc, distfun=distfunc,
          RowSideColors=coll, 
          ColSideColors=col2,
          Colv=T,Rowv=T,key=T,
          col=col,
          cexCol=0.65,
          keysize=1,
          labRow=NA,
          trace="none",
          density.info="none")
legend=unique(data.frame(col2,cl.col$labels))
legend("right",legend=legend[,2],fill=legend[,1])
? legend
dev.off()
```

The following figure is almost same with the above one, except I remove the RowSideColors, I just want to show the number of the gene I used.(ok. choose this one to the paper). By the way, no need to do that in windows, you can not get better figure even in the windows. 
```{r}
library("gplots")
library("RColorBrewer")
library("grDevices")
mydata<-f2[which(gsi_2[,3]>0.6),]
save(mydata,file="GWBS.mhl.heatmap.RData")
hclustfunc <- function(x) hclust(x, method="ward.D")
distfunc <- function(x) dist(x, method="euclidean")
# perform clustering on rows and columns
cl.row <- hclustfunc(distfunc(mydata))
cl.col <- hclustfunc(distfunc(t(mydata)))
# extract cluster assignments; i.e. k=8 (rows) k=5 (columns), ncut should smaller than total color number. "Set1" of brewer.pal <12
ncut=6  
gr.row <- cutree(cl.row, ncut)
# gr.col <- cutree(cl.col, 5)
require(RColorBrewer)
col1 <- brewer.pal(ncut, "Set1")[gr.row[cl.row$order]]     # the maximum of brewer.pal is 12
#col2 <- brewer.pal(5, "Pastel1")

c25 <- c("dodgerblue2","#E31A1C", # red
         "green4",
         "#6A3D9A", # purple
         "#FF7F00", # orange
         "black","gold1",
         "skyblue2","#FB9A99", # lt pink
         "palegreen2",
         "#CAB2D6", # lt purple
         "#FDBF6F", # lt orange
         "gray70", "khaki2",
         "maroon","orchid1","deeppink1","blue1","steelblue4",
         "darkturquoise","green1","yellow4","yellow3",
         "darkorange4","brown")

coll<-gr.row[as.numeric(as.factor(cl.row$labels))]
col2<-c25[as.numeric(as.factor(cl.col$labels))]
library("grDevices")
library("gplots")
mydata[mydata<0]<-0
mydata[mydata>1]<-1
col=colorRampPalette(c("blue", "yellow"))(20) 
save(mydata,file="wgbs.mhl.heatmap.RData")
pdf("heatmap.wgbs.mhl.gsi0.6.with.H1.wbc.cancer.pdf")
heatmap.2(as.matrix((mydata)),hclustfun=hclustfunc, distfun=distfunc,
          ColSideColors=col2,
          Colv=T,Rowv=T,key=T,
          col=col,
          cexCol=0.65,
          keysize=1,
          labRow=NA,
          margins=c(5,10),
          trace="none",
          srtCol=45,
          density.info="none")
legend=unique(data.frame(col2,cl.col$labels))
legend(x=0.85,y=0.8,legend=legend[,2],col=as.character(legend[,1]),pch=15,cex = 0.5)
dev.off()
```

Now we try to analysis the correlation between adjacent CpGs in methylation haplotype block regions were the distance between two CpGs. 

```{r,echo=TRUE}
setwd("/home/shg047/monod/r2gap")
load("STL003SG-01.all_chrs.hapInfo.txt.pairwiseR2.gapR2.RData")

PairwiseR2plot <- function(R2file=args[1],samplingsize=5000){
  file=R2file
  library(RColorBrewer)
  require(KernSmooth)
  library("fields")
  
  fudgeit <- function(){
    xm <- get('xm', envir = parent.frame(1))
    ym <- get('ym', envir = parent.frame(1))
    z  <- get('dens', envir = parent.frame(1))
    colramp <- get('colramp', parent.frame(1))
    image.plot(xm,ym,z, col = colramp(256), legend.only = T, add =F)
  }
  
  my.cols <- rev(brewer.pal(11, "RdYlBu"))
  data<-read.table(file,sep="\t",skip=1)
  id<-unique(data[,1])
  gap1<-c()
  gap2<-c()
  r2<-c()
  
  for(j in 1:length(id)){
    tmp<-data[data[,1]==id[j],]
    tmp<-na.omit(tmp)
    if(nrow(tmp)>2 && sd(tmp[,3],na.rm=T) != 0 && summary(lm(tmp[,3]~tmp[,2],na.action="na.omit"))$coefficients[2,1]<0){
      r2<-c(r2,tmp[,3])
      gap1<-c(gap1,tmp[,2])
      gap2<-c(gap2,tmp[,2]/max(tmp[,2],na.rm=T))
      if (length(gap1)>samplingsize) break
      print(c(j,length(gap1)))
    }
  }
  rlt<-data.frame(gap1,gap2,r2)
  save(rlt,file=paste(file,".gapR2.RData",sep=""))
  pdf(paste(file,".paisewise.R2-2.smooth.pdf",sep=""))
  par(mar = c(5,4,4,5) + .1)
  par(mfrow=c(2,2))
  smoothScatter(x=gap1,y=r2, nrpoints=.3*length(gap1), colramp=colorRampPalette(my.cols), pch=19, cex=.3, col = "green1",xlim=c(0,150),postPlotHook = fudgeit)
  smoothScatter(x=gap2,y=r2, nrpoints=.3*length(gap2), colramp=colorRampPalette(my.cols), pch=19, cex=.3, col = "green1",xlim=c(0,150),postPlotHook = fudgeit)
  plot(r2~gap1)
  plot(r2~gap2)
  dev.off()
}
```

This script were used to add legend for smoothscatter plot. 
```{r}

library(RColorBrewer)
require(KernSmooth)
library("fields")
fudgeit <- function(){
  xm <- get('xm', envir = parent.frame(1))
  ym <- get('ym', envir = parent.frame(1))
  z  <- get('dens', envir = parent.frame(1))
  colramp <- get('colramp', parent.frame(1))
  image.plot(xm,ym,z, col = colramp(256), legend.only = T, add =F)
}

file=list.files(pattern="*gapR2.RData$")
my.cols <- rev(brewer.pal(11, "RdYlBu"))

for(i in 1:length(file)){
  load(file[i])
  gap1=rlt[,1]
  gap2=rlt[,2]
  r2=rlt[,3]
  
  output=paste(file[i],".paisewise.R2-2.smooth.pdf",sep="")
  pdf(output)
  par(mar = c(5,4,4,3) + .1)
  par(mfrow=c(2,2))
  smoothScatter(x=gap1,y=r2, nrpoints=.3*length(gap1), colramp=colorRampPalette(my.cols), pch=19, cex=.3, col = "green1",xlim=c(0,150),postPlotHook = fudgeit)
  smoothScatter(x=gap2,y=r2, nrpoints=.3*length(gap2), colramp=colorRampPalette(my.cols), pch=19, cex=.3, col = "green1",xlim=c(0,1),postPlotHook = fudgeit)
  plot(r2~gap1)
  plot(r2~gap2)
  dev.off()
  print(output)
}
```

Here, I try to combine RRBS, GW-BS data and select some biomarkers for cancer diagnosis and tissue mapping.
first, I need to collect the seqencing hot splot by RRBS and GWBS so that these biomarkes can be staible in further studies.
second, we need to plot the distribution of the GSI, in the WB-seq and RRBS dataset


```{r}

d<-read.table("group.txt",head=T,sep="\t",as.is=T)
pdf("pie.pdf")
pie(as.numeric(d[1,]),col=rainbow(length(d)),labels=names(d),cex=0.5)
dev.off()

setwd("/home/shg047/monod/methyblock")
name1<-c("Intergenic","Downstream","Enhancer","Exon","Intron","Promoter","UTR3","UTR5","miRNA")
q1<-c(118131,10847,9006,50517,104648,129263,5998,57108,468)
name2<-c("CpG Island","CpG Shore","CpG Shelf")
q2<-c(79704,26098,3243)
names(q1)=name1
names(q2)=name2

q3<-q1/sum(q1)
q4<-q2/sum(q2)
names(q3)=name1
names(q4)=name2

pdf("distribution1a.pdf")
par(mfrow=c(2,2),mar=c(1,2,3,2))
pie(q1,col=rainbow(length(q1)),lwd=2,radius=0.8,cex=1)
pie(q2,col=rainbow(length(q2)),lwd=2,radius=0.8,cex=1)
par(mar=c(5,6,1,3))
bpt1<-barplot(q1,col=rainbow(length(q3)),lwd=2,horiz=T,las=2)
text(x= q1+3000, y= bpt1, labels=as.character(q1), xpd=TRUE)
bpt2<-barplot(q2,col=rainbow(length(q4)),lwd=2,horiz=T,las=2)
text(x= q2+200, y= bpt2, labels=as.character(q2), xpd=TRUE)
dev.off()

pdf("distribution1b.pdf")
par(mfrow=c(2,2),mar=c(1,2,3,2))
pie(q1,col=rainbow(length(q1)),lwd=2,radius=0.8,cex=1)
pie(q2,col=rainbow(length(q2)),lwd=2,radius=0.8,cex=1)
par(mar=c(5,6,1,3))
bpt1<-barplot(q3,col=rainbow(length(q3)),lwd=2,horiz=T,las=2)
text(x= q3+0.1, y= bpt1, labels=round(q3,3), xpd=TRUE)
bpt2<-barplot(q4,col=rainbow(length(q4)),lwd=2,horiz=T,las=2)
text(x= q4+0.15, y= bpt2, labels=round(q4,3), xpd=TRUE)
dev.off()

```


Here, we try to merge RRBS and GWBS data and find biomarker
```{r}
setwd("/home/shg047/monod/rrbs_kun")
data<-read.table("rrbs.kun.mhl.txt",head=T,sep="\t",as.is=T,row.names=1,check.names=F)
newdata<-RawNARemove(data,missratio=0.45)
newdata<-ColNARemove(newdata,missratio=0.5)
dim(data)
dim(newdata)
newdata<-impute.knn(data.matrix(newdata))$data
colnames(newdata)
length(grep("6-P",colnames(newdata)))
length(grep("7-P",colnames(newdata)))
length(grep("PC-P",colnames(newdata)))
length(grep("NC-P",colnames(newdata)))
length(grep("6-T",colnames(newdata)))
length(grep("7-T",colnames(newdata)))
length(grep("PC-T",colnames(newdata)))
length(grep("RRBS-6P",colnames(newdata)))
length(grep("RRBS-7P",colnames(newdata)))
# all kinds of  biomarker and tissue specifal regions
write.table(newdata,file="rrbs.kun.mhl.na.impute.txt",col.names=NA,row.names=T,quote=F)

```

hotspot region for RRBS encode. 80% hot-regions were collected. merge rrbs-encode-hotspot with WGBS and RRBS_kun regions

```{r}
# Come to UT server
# RRBS encode project
# setwd("/home/shg047/monod/rrbs_encode")
setwd("/home/sguo/monod/rrbs")
load("encode.rrbs.pos.freq.RData")
pdf("RRBS.Region.Frequency.pdf")
hist(encode.rrbs.pos.freq,breaks=100,col="red",ylim=c(0,500000),xlab="Counts of Samples")
dev.off()
rrbs_hot=encode.rrbs.pos.freq[encode.rrbs.pos.freq>50]
bed1<-cor2bed(names(rrbs_hot))

# RRBS_kun Regions
setwd("/home/sguo/monod/rrbs_kun")
setwd("/home/shg047/work/monod/rrbs_kun")
newdata1<-read.table("rrbs.kun.mhl.na.impute.txt",head=T,as.is=T,row.names=1,check.names=F)
bed2<-cor2bed(rownames(newdata1))
dim(newdata1)
dim(bed2)
head(newdata1)

# WGBS regions
library("impute")
library("preprocessCore")
library("sva")
setwd("/home/sguo/monod/hap/wgbs/All_chromosomes_combined")
newdata2<-read.table("wgbs.mhl.txt",head=T,sep="\t",as.is=T,row.names=1,check.names=F)
file1<-newdata2
f2<-RawNARemove(file1,missratio=0.3)
f2<-impute.knn(data.matrix(f2))$data
colnames(f2) 
colnames(f2)<-gsub("_","-",colnames(f2))
colname2<-unlist(lapply(colnames(f2),function(x) unlist(strsplit(x,"[.]"))[1]))
colname2
colnames(f2)<-colname2
saminfo2<-read.table("/home/sguo/monod/hap/wgbs/All_chromosomes_combined/newsaminfo.txt",head=T,sep="\t",as.is=T)
saminfo2<-saminfo2[match(colname2,saminfo2[,1]),]
saminfo2
colnames(f2)<-saminfo2[,2]
colnames(f2)
newdata2<-f2
bed3<-cor2bed(rownames(newdata2))
# do the tissue specific mhl regions
# total tissue specific regions
# colon, lung, pancreatic specific regions
ob1<-grep("Lung",colnames(newdata2))
ob2<-grep("Pancreas",colnames(newdata2))
ob3<-grep("Colon",colnames(newdata2))
ob<-c(ob1,ob2,ob3)
gsidata<-newdata2[,ob]
hgsi<-gsi(gsidata)
Lung<-subset(hgsi,group="Lung")[order(subset(hgsi,group="Lung")[,3],decreasing=T)[1:400],]
Colon<-subset(hgsi,group="Colon")[order(subset(hgsi,group="Colon")[,3],decreasing=T)[1:400],]
Pancreas<-subset(hgsi,group="Pancreas")[order(subset(hgsi,group="Pancreas")[,3],decreasing=T)[1:400],]
hgsi_sub<-rbind(Lung,Colon,Pancreas)
bed3<-cor2bed(hgsi_sub[,1])


# merge the data
head(bed1)  # rrbs_encode
head(bed2)  # rrbs_kun
head(bed3)  # gwbs
dim(bed1)
dim(bed2)
dim(bed3)

# bedtools intersection
bed21<-Rbedtools("intersectBed",bed2,bed1,opt.string="-wa -u")
bed213<-Rbedtools("intersectBed",bed21,bed3,opt.string="-wa -u")
dim(bed21)
dim(bed213)

cor<-bed2cor(bed213)
newdata<-newdata1[match(cor,rownames(newdata1)),]
dim(newdata)
# built prediction model
wgbs_new<-newdata2[match(cor,rownames(newdata2)),]
rrbs_new<-newdata1[match(cor,rownames(newdata1)),]
head(wgbs_new)
head(rrbs_new)
dim(wgbs_new)
dim(rrbs_new)

data<-cbind(wgbs_new,rrbs_new)

ob1<-grep("Lung",colnames(data))
ob2<-grep("Pancreas",colnames(data))
ob3<-grep("Colon",colnames(data))
ob<-c(ob1,ob2,ob3)
rfdata<-data[,(1:61)[ob]]
ylab<-colnames(data)[(1:61)[ob]]

# batch 1 plasma mapping
ob1<-grep("6-P",colnames(data))
ob2<-grep("7-P",colnames(data))
ob3<-grep("PC-P",colnames(data))
ob4<-grep("RRBS-6P",colnames(data))
ob5<-grep("RRBS-7P",colnames(data))

pred1<-data[,c(ob1,ob2,ob3)]
lab1<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Pancreas",length(ob3))))

save(data,file="GWBS.RRBS.67PC.GSI.MHL.RData")

library("randomForest")
bestmtry<-tuneRF(x=t(pred1),y=lab1,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
rf.model1<-randomForest(x=t(pred1),y=lab1,mtry=best.mtry)
rf.model1
rf.model1<-randomForest(x=t(rfdata),y=as.factor(ylab),xtest=t(pred1),ytest=lab1,mtry=best.mtry)
rf.model1
```

# 6688-> 3781 
# merge tissue specific regions (3781 cross with)
# scp shg047@genome-miner.ucsd.edu:/home/shg047/monod/dec/Table.GSI.WGBS.Remove.H1.WBC.rlt.txt ./
# scp shg047@genome-miner.ucsd.edu:/home/shg047/monod/dec/Table.GSI.WGBS.with.H1.WBC.rlt.txt ./
# tune the number of the tissue specific regions is very important

```{r}
# Come to UT server
# RRBS encode project
# setwd("/home/shg047/monod/rrbs_encode")
setwd("/home/sguo/monod/rrbs")
load("encode.rrbs.pos.freq.RData")
pdf("RRBS.Region.Frequency.pdf")
hist(encode.rrbs.pos.freq,breaks=100,col="red",ylim=c(0,500000),xlab="Counts of Samples")
dev.off()
rrbs_hot=encode.rrbs.pos.freq[encode.rrbs.pos.freq>50]
bed1<-cor2bed(names(rrbs_hot))

# RRBS_kun Regions
# setwd("/home/shg047/monod/rrbs_kun")
setwd("/home/sguo/monod/rrbs_kun")
newdata1<-read.table("rrbs.kun.mhl.na.impute.txt",head=T,as.is=T,row.names=1,check.names=F)
bed2<-cor2bed(rownames(newdata1))
dim(newdata1)
dim(bed2)
head(newdata1)

# WGBS regions
library("impute")
library("preprocessCore")
library("sva")
setwd("/home/sguo/monod/hap/wgbs/All_chromosomes_combined")
newdata2<-read.table("wgbs.mhl.txt",head=T,sep="\t",as.is=T,row.names=1,check.names=F)
file1<-newdata2
f2<-RawNARemove(file1,missratio=0.3)
f2<-impute.knn(data.matrix(f2))$data
# f2.t1<-normalize.quantiles(f2[,13:58])
# batch=c(rep(1,10),rep(2,36))
# f2.t2<-ComBat(f2.t1, batch, mod=NULL, par.prior = TRUE,prior.plots = FALSE)
# f2[,13:58]<-f2.t2
# re-assign colnames
colnames(f2) 
colnames(f2)<-gsub("_","-",colnames(f2))
colname2<-unlist(lapply(colnames(f2),function(x) unlist(strsplit(x,"[.]"))[1]))
colname2
colnames(f2)<-colname2
saminfo2<-read.table("/home/sguo/monod/hap/wgbs/All_chromosomes_combined/newsaminfo.txt",head=T,sep="\t",as.is=T)
saminfo2<-saminfo2[match(colname2,saminfo2[,1]),]
saminfo2
colnames(f2)<-saminfo2[,2]
colnames(f2)
newdata2<-f2
bed3<-cor2bed(rownames(newdata2))

bed21<-Rbedtools("intersectBed",bed2,bed1,opt.string="-wa -u")
bed213<-Rbedtools("intersectBed",bed21,bed3,opt.string="-wa -u")
dim(bed21)
dim(bed213)
cor<-bed2cor(bed213)

setwd("/home/sguo/monod/hap/wgbs/All_chromosomes_combined")
tsr1<-read.table("Table.GSI.WGBS.Remove.H1.WBC.rlt.txt",sep="\t",as.is=T,row.names=1,head=T)
tsr2<-read.table("Table.GSI.WGBS.with.H1.WBC.rlt.txt",sep="\t",as.is=T,row.names=1,head=T)

bed41<-cor2bed(as.character(rownames(tsr1)))
bed42<-cor2bed(as.character(rownames(tsr2)))

bed21341<-Rbedtools("intersectBed",bed213,bed41,opt.string="-wa -u")
bed21342<-Rbedtools("intersectBed",bed213,bed42,opt.string="-wa -u")

dim(bed21341)
dim(bed21342)
# 
cor<-bed2cor(bed21341)
newdata<-newdata1[match(cor,rownames(newdata1)),]
dim(newdata)
# built prediction model

cor21341<-bed2cor(bed21341)
wgbs_new<-newdata2[match(cor21341,rownames(newdata2)),]
rrbs_new<-newdata1[match(cor21341,rownames(newdata1)),]
head(wgbs_new)
head(rrbs_new)
dim(wgbs_new)
dim(rrbs_new)

data<-cbind(wgbs_new,rrbs_new)
write.table(data,file="WGBS.RRBS.Batch1plus2.na.remove.26Dec2015.txt",col.names=NA,row.names=T,quote=F,sep="\t")

# classification to all the tissue is very bad
exc1<-grep("H1",colnames(data))
exc2<-grep("WB",colnames(data))
exc3<-grep("CCT",colnames(data))
exc4<-grep("HCT",colnames(data))
exc<-sort(c(exc1,exc2,exc3,exc4))

rfdata<-data[,(1:61)[-exc]]
ylab<-colnames(data)[(1:61)[-exc]]
table(ylab)
library("randomForest")
# Don't do random forest in GWBS dataset. the performance is very bad
# bestmtry<-tuneRF(x=t(rfdata), y=as.factor(ylab), ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
# best.mtry=bestmtry$mtry
# rf.model1<-randomForest(t(rfdata),as.factor(ylab),mtry=best.mtry,mtrees=mtrees)
# so only extract lung, colon, pancreatic and do the prediction again
ob1<-grep("Lung",colnames(data))
ob2<-grep("Pancreas",colnames(data))
ob3<-grep("Colon",colnames(data))
ob<-c(ob1,ob2,ob3)
rfdata<-data[,(1:61)[ob]]
ylab<-colnames(data)[(1:61)[ob]]
ylab
bestmtry<-tuneRF(x=t(rfdata), y=as.factor(ylab), ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
rf.model1<-randomForest(t(rfdata),as.factor(ylab),mtry=best.mtry,mtrees=mtrees)

# RRBS Batch 1 and Batch 2
ob1<-grep("6-P",colnames(data))
ob2<-grep("7-P",colnames(data))
ob3<-grep("PC-P",colnames(data))
ob4<-grep("RRBS-6P",colnames(data))
ob5<-grep("RRBS-7P",colnames(data))

pred1<-data[,c(ob1,ob2,ob3,ob4,ob5)]
head(pred1)
lab1<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Pancreas",length(ob3)),rep("Colon",length(ob4)),rep("Lung",length(ob5))))
colnames(pred1)<-lab1
hgsi1<-gsi(data.matrix(pred1))
pred2<-pred1[match(subset(hgsi1,GSI>0.4)[,1],rownames(pred1)),]
head(pred2)
lab2<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Pancreas",length(ob3)),rep("Colon",length(ob4)),rep("Lung",length(ob5))))

library("randomForest")
bestmtry<-tuneRF(x=t(pred2),y=lab2,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
rf.model1<-randomForest(x=t(pred2),y=lab2,mtry=best.mtry)
rf.model1





# built tissue mapping model

```

# back-up data
# WGBS MHL matrix(update by Shicheng) were save in (UTH):
/home/sguo/monod/hap/wgbs/All_chromosomes_combined/wgbs.mhl.txt
# RRBS(Batch 1 and 2) MHL matrix(update by Shicheng) were save in (UTH):
/home/sguo/monod/rrbs_kun/rrbs.kun.mhl.txt

Here, I use the Batch 1 and batch 2 data do the random forest directly without any GSI filter. 
```{r}
# RRBS_kun Regions
# setwd("/home/shg047/monod/rrbs_kun")
setwd("/home/sguo/monod/rrbs_kun")
newdata1<-read.table("rrbs.kun.mhl.na.impute.txt",head=T,as.is=T,row.names=1,check.names=F)
bed2<-cor2bed(rownames(newdata1))
dim(newdata1)
dim(bed2)
head(newdata1)

# batch 1
data<-newdata1
ob1<-grep("6-P",colnames(data))
ob2<-grep("7-P",colnames(data))
ob3<-grep("PC-P",colnames(data))

pred1<-data[,c(ob1,ob2,ob3)]
head(pred1)
lab1<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Pancreas",length(ob3))))
colnames(pred1)<-lab1
hgsi1<-gsi(data.matrix(pred1))
pred2<-pred1[match(subset(hgsi1,GSI>0.4)[,1],rownames(pred1)),]
head(pred2)
lab2<-lab1
dim(pred2)

library("randomForest")
bestmtry<-tuneRF(x=t(pred2),y=lab2,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
rf.model1<-randomForest(x=t(pred2),y=lab2,mtry=best.mtry)
rf.model1

# batch 2

ob4<-grep("RRBS-6P",colnames(data))
ob5<-grep("RRBS-7P",colnames(data))

pred1<-data[,c(ob4,ob5)]
head(pred1)
lab1<-as.factor(c(rep("Colon",length(ob4)),rep("Lung",length(ob5))))
colnames(pred1)<-lab1
hgsi1<-gsi(data.matrix(pred1))
pred2<-pred1[match(subset(hgsi1,GSI>0.4)[,1],rownames(pred1)),]
head(pred2)
lab2<-lab1
dim(pred2)

library("randomForest")
bestmtry<-tuneRF(x=t(pred2),y=lab2,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
rf.model1<-randomForest(x=t(pred2),y=lab2,mtry=best.mtry)
rf.model1

# batch 1 and batch 2 together
data<-newdata1
ob1<-grep("6-P",colnames(data))
ob2<-grep("7-P",colnames(data))
ob3<-grep("PC-P",colnames(data))
ob4<-grep("RRBS-6P",colnames(data))
ob5<-grep("RRBS-7P",colnames(data))

pred1<-data[,c(ob1,ob2,ob3,ob4,ob5)]
head(pred1)
lab1<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Pancreas",length(ob3)),rep("Colon",length(ob4)),rep("Lung",length(ob5))))
colnames(pred1)<-lab1
hgsi1<-gsi(data.matrix(pred1))
pred2<-pred1[match(subset(hgsi1,GSI>0.4)[,1],rownames(pred1)),]
head(pred2)
lab2<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Pancreas",length(ob3)),rep("Colon",length(ob4)),rep("Lung",length(ob5))))
dim(pred2)

library("randomForest")
bestmtry<-tuneRF(x=t(pred2),y=lab2,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
rf.model1<-randomForest(x=t(pred2),y=lab2,mtry=best.mtry)
rf.model1
```

Great. the prediction were beautiful when I do it direclty with RRBS_kun's data. Now, I am trying to overlap with GWBS and GSI(0.2 or 0.3), don't use too large GSI.

```{r}
# RRBS_Encode
setwd("/home/sguo/monod/rrbs")
load("encode.rrbs.pos.freq.RData")
rrbs_hot=encode.rrbs.pos.freq[encode.rrbs.pos.freq>50]
bed1<-cor2bed(names(rrbs_hot))

# RRBS_kun Regions
setwd("/home/shg047/monod/rrbs_kun")
setwd("/home/sguo/monod/rrbs_kun")
newdata1<-read.table("rrbs.kun.mhl.na.impute.txt",head=T,as.is=T,row.names=1,check.names=F)
bed2<-cor2bed(rownames(newdata1))
dim(newdata1)
dim(bed2)
head(newdata1)
# WGBS regions
library("impute")
library("preprocessCore")
library("sva")
setwd("/home/sguo/monod/hap/wgbs/All_chromosomes_combined")
newdata2<-read.table("wgbs.mhl.txt",head=T,sep="\t",as.is=T,row.names=1,check.names=F)
file1<-newdata2
f2<-RawNARemove(file1,missratio=0.3)
f2<-impute.knn(data.matrix(f2))$data
# f2.t1<-normalize.quantiles(f2[,13:58])
# batch=c(rep(1,10),rep(2,36))
# f2.t2<-ComBat(f2.t1, batch, mod=NULL, par.prior = TRUE,prior.plots = FALSE)
# f2[,13:58]<-f2.t2
# re-assign colnames
colnames(f2) 
colnames(f2)<-gsub("_","-",colnames(f2))
colname2<-unlist(lapply(colnames(f2),function(x) unlist(strsplit(x,"[.]"))[1]))
colname2
colnames(f2)<-colname2
saminfo2<-read.table("/home/sguo/monod/hap/wgbs/All_chromosomes_combined/newsaminfo.txt",head=T,sep="\t",as.is=T)
saminfo2<-saminfo2[match(colname2,saminfo2[,1]),]
saminfo2
colnames(f2)<-saminfo2[,2]
colnames(f2)
newdata2<-f2
exc1<-grep("H1",colnames(newdata2))
exc2<-grep("WB",colnames(newdata2))
exc3<-grep("CCT",colnames(newdata2))
exc4<-grep("HCT",colnames(newdata2))
exc<-sort(c(exc1,exc2,exc3,exc4))
gsidata<-newdata2[,(1:61)[-exc]]
dim(gsidata)
colnames(gsidata)
hgsi<-gsi(gsidata)
hgsi_sub<-subset(hgsi,GSI>0.3)
bed3<-cor2bed(hgsi_sub[,1])

# merge the data
bed21<-Rbedtools("intersectBed",bed2,bed1,opt.string="-wa -u")
bed213<-Rbedtools("intersectBed",bed21,bed3,opt.string="-wa -u")
dim(bed21)
dim(bed213)
cor<-bed2cor(bed213)

newdata<-newdata1[match(cor,rownames(newdata1)),]
dim(newdata)
colnames(newdata)
data<-newdata
# bilud the prediction model
# batch 1
data<-newdata1
ob1<-grep("6-P",colnames(data))
ob2<-grep("7-P",colnames(data))
ob3<-grep("PC-P",colnames(data))
ob4<-grep("NC-P",colnames(data))[1:10]
ob5<-grep("NC-P",colnames(data))[11:20]

pred1<-data[,c(ob1,ob2,ob3,ob4,ob5)]
head(pred1)
lab1<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Pancreas",length(ob3)),rep("Normal",length(ob4)),rep("Normal",length(ob5))))
colnames(pred1)<-lab1
hgsi1<-gsi(data.matrix(pred1))
pred2<-pred1[match(subset(hgsi1,GSI>0.4)[,1],rownames(pred1)),]
batch1_feature<-pred2

head(pred2)
lab2<-lab1<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Pancreas",length(ob3)),rep("Normal1",length(ob4)),rep("Normal2",length(ob5))))

dim(pred2)


library("randomForest")
x<-matrix(0,ncol=6,nrow=5)
for(i in 1:100){
  bestmtry<-tuneRF(x=t(pred2),y=lab2,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
  best.mtry=bestmtry$mtry
  rf.model1<-randomForest(x=t(pred2),y=lab2,mtry=best.mtry)
  x<-rf.model1$confusion+x
}
x<-x/100
write.table(x,file="random.forest.confusion.batch1.RRBS.txt",sep="\t",quote=F,col.names=NA,row.names=T)

# batch 2
ob4<-grep("RRBS-6P",colnames(data))
ob5<-grep("RRBS-7P",colnames(data))

pred1<-data[,c(ob4,ob5)]
head(pred1)
lab1<-as.factor(c(rep("Colon",length(ob4)),rep("Lung",length(ob5))))
colnames(pred1)<-lab1
hgsi1<-gsi(data.matrix(pred1))
pred2<-pred1[match(subset(hgsi1,GSI>0.4)[,1],rownames(pred1)),]
batch2_feature<-pred2

head(pred2)
lab2<-lab1
dim(pred2)

library("randomForest")
x<-matrix(0,ncol=3,nrow=2)
bestmtry<-tuneRF(x=t(pred2),y=lab2,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
for(i in 1:100){
  rf.model1<-randomForest(x=t(pred2),y=lab2,mtry=best.mtry)
  x<-rf.model1$confusion+x
}
x<-x/100

write.table(x,file="random.forest.confusion.batch2.RRBS.txt",sep="\t",quote=F,col.names=NA,row.names=T)


# feature comparsion
match(rownames(batch1_feature), rownames(batch2_feature))

# batch 1 and batch 2 together
data<-newdata1
ob1<-grep("6-P",colnames(data))
ob2<-grep("7-P",colnames(data))
# ob3<-grep("PC-P",colnames(data))
ob4<-grep("RRBS-6P",colnames(data))
ob5<-grep("RRBS-7P",colnames(data))
ob6<-grep("NC-P",colnames(data))

pred1<-data[,c(ob1,ob2,ob4,ob5,ob6)]
head(pred1)
lab1<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Colon",length(ob4)),rep("Lung",length(ob5)),rep("Normal",length(ob6))))
colnames(pred1)<-lab1
hgsi1<-gsi(data.matrix(pred1))
pred2<-pred1[match(subset(hgsi1,GSI>0.44)[,1],rownames(pred1)),]
batch12_feature<-pred2
write.table(batch12_feature,file="random.forest.confusion.batch1.and.2.RRBS.InputMatrix.txt",sep="\t",quote=F,col.names=NA,row.names=T)
head(pred2)
lab2<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Colon",length(ob4)),rep("Lung",length(ob5)),rep("Normal",length(ob6))))
dim(pred2)

library("randomForest")
x<-matrix(0,ncol=4,nrow=3)
bestmtry<-tuneRF(x=t(pred2),y=lab2,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
for(i in 1:100){
  rf.model1<-randomForest(x=t(pred2),y=lab2,mtry=best.mtry)
  x<-rf.model1$confusion+x
}
x=x/100
write.table(x,file="random.forest.confusion.batch1.and.2.RRBS.txt",sep="\t",quote=F,col.names=NA,row.names=T)

```

Now, we need to identify some cancer biomarker through comparion between tissue and normal. 
```{r}
setwd("/home/shg047/monod/rrbs_kun")
setwd("/home/sguo/monod/rrbs_kun")
newdata1<-read.table("rrbs.kun.mhl.na.impute.txt",head=T,as.is=T,row.names=1,check.names=F)
bed2<-cor2bed(rownames(newdata1))
dim(newdata1)
dim(bed2)
head(newdata1)
data<-newdata1
ob1<-grep("6-P",colnames(data))
ob2<-grep("7-P",colnames(data))
ob3<-grep("PC-P",colnames(data))
ob4<-grep("RRBS-6P",colnames(data))
ob5<-grep("RRBS-7P",colnames(data))
ob6<-grep("NC-P",colnames(data))

length(ob1)+length(ob4)
data_colon<-data[,c(ob1,ob4,ob6)]
data_lung<-data[,c(ob2,ob5,ob6)]
data_pancrease<-data[,c(ob3,ob6)]

p1<-ttestPValue(data_colon,1:30,31:50)
p1.fdr<-p.adjust(p2[,2],"fdr")
sum(p1.fdr<0.05)
pred2=data_colon[which(p1.fdr<0.05),]
lab2<-c(rep("lung",30),rep("normal",20))
bestmtry<-tuneRF(x=t(pred2),y=lab2,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
rf.model1<-randomForest(x=t(pred2),y=lab2,mtry=best.mtry)
x<-rf.model1$confusion+x




write.table(p1,file="colon.dmr.fdr0.5.txt",col.names=NA,row.names=T,sep="\t")

p2<-ttestPValue(data_lung,1:29,30:49)
p2.fdr<-p.adjust(p2[,2],"fdr")
sum(p2.fdr<0.05)
write.table(p2,file="lung.dmr.fdr0.5.txt",col.names=NA,row.names=T,sep="\t")

p3<-ttestPValue(data_pancrease,1:10,11:30)
p3.fdr<-p.adjust(p3[,2],"fdr")
sum(p3.fdr<0.05)
write.table(p3,file="pancrease.dmr.fdr0.5.txt",col.names=NA,row.names=T,sep="\t")



```


Here, Dr. Zhang don't want to merge RRBS data from ENCODE Project. He just want to shared regions from our RRBS and GWBS data
```{r}
# RRBS_kun Regions
setwd("/home/shg047/monod/rrbs_kun")
setwd("/home/sguo/monod/rrbs_kun")
newdata1<-read.table("rrbs.kun.mhl.na.impute.txt",head=T,as.is=T,row.names=1,check.names=F)
bed2<-cor2bed(rownames(newdata1))
dim(newdata1)
dim(bed2)
head(newdata1)
# WGBS regions
library("impute")
library("preprocessCore")
library("sva")
setwd("/home/shg047/monod/hap/wgbs/All_chromosomes_combined/")
setwd("/home/sguo/monod/hap/wgbs/All_chromosomes_combined")
newdata2<-read.table("wgbs.mhl.txt",head=T,sep="\t",as.is=T,row.names=1,check.names=F)
file1<-newdata2
f2<-RawNARemove(file1,missratio=0.3)
f2<-impute.knn(data.matrix(f2))$data
# f2.t1<-normalize.quantiles(f2[,13:58])
# batch=c(rep(1,10),rep(2,36))
# f2.t2<-ComBat(f2.t1, batch, mod=NULL, par.prior = TRUE,prior.plots = FALSE)
# f2[,13:58]<-f2.t2
# re-assign colnames
colnames(f2) 
colnames(f2)<-gsub("_","-",colnames(f2))
colname2<-unlist(lapply(colnames(f2),function(x) unlist(strsplit(x,"[.]"))[1]))
colname2
colnames(f2)<-colname2
saminfo2<-read.table("/home/shg047/monod/hap/wgbs/All_chromosomes_combined/newsaminfo.txt",head=T,sep="\t",as.is=T)
saminfo2<-read.table("/home/sguo/monod/hap/wgbs/All_chromosomes_combined/newsaminfo.txt",head=T,sep="\t",as.is=T)

saminfo2<-saminfo2[match(colname2,saminfo2[,1]),]
saminfo2
colnames(f2)<-saminfo2[,2]
colnames(f2)
newdata2<-f2
exc1<-grep("H1",colnames(newdata2))
# exc2<-grep("WB",colnames(newdata2))
exc3<-grep("CCT",colnames(newdata2))
exc4<-grep("HCT",colnames(newdata2))
exc5<-grep("Bladder|Thymus|Ovary|Fat cell|Vessel|Heart|muscle",colnames(newdata2))
exc<-sort(c(exc1,exc2,exc3,exc4,exc5))
gsidata<-newdata2[,(1:61)[-exc]]
dim(gsidata)
colnames(gsidata)
hgsi<-gsi(gsidata)
hgsi_sub<-subset(hgsi,GSI>0.5)
bed3<-cor2bed(hgsi_sub[,1])
dim(gsidata)
dim(hgsi_sub)

# merge the data
bed23<-Rbedtools("intersectBed",bed2,bed3,opt.string="-wa -u")
dim(bed23)
cor<-bed2cor(bed23)

newdata<-newdata1[match(cor,rownames(newdata1)),]
dim(newdata)
colnames(newdata)
data<-newdata
# bilud the prediction model
# batch 1
ob1<-grep("6-P",colnames(data))
ob2<-grep("7-P",colnames(data))
ob3<-grep("PC-P",colnames(data))
ob4<-grep("NC-P",colnames(data))[1:10]
ob5<-grep("NC-P",colnames(data))[11:20]

pred1<-data[,c(ob1,ob2,ob3,ob4,ob5)]
dim(pred1)
lab1<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Pancreas",length(ob3)),rep("Normal",length(ob4)),rep("Normal",length(ob5))))
colnames(pred1)<-lab1
hgsi1<-gsi(data.matrix(pred1))
pred2<-pred1[match(subset(hgsi1,GSI>0.4)[,1],rownames(pred1)),]
batch1_feature<-pred2

head(pred2)
lab2<-lab1<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Pancreas",length(ob3)),rep("Normal1",length(ob4)),rep("Normal2",length(ob5))))
dim(pred2)
rf.importance<-rep(0,nrow(pred2))
library("randomForest")
x<-matrix(0,ncol=6,nrow=5)
sensitivity<-c()
specificity<-c()

bestmtry<-tuneRF(x=t(pred2),y=lab2,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry

for(i in 1:100){
rf.model1<-randomForest(x=t(pred2),y=lab2,mtry=best.mtry)
rf.importance<-rf.importance+rf.model1$importance/100
x<-rf.model1$confusion+x
x1<-x
x1[,3]<-x1[,3]+x1[,4]
x1<-x1[,-4]
x1[3,]<-x1[3,]+x1[4,]
x1<-x1[-4,]
sen<-c(x1[1,1]/sum(x1[1,]),x1[2,2]/sum(x1[2,]),x1[3,3]/sum(x1[3,]),x1[4,4]/sum(x1[4,]))
spe<-c(1-sum(x1[-1,1])/sum(x1[-1,]),1-sum(x1[-2,2])/sum(x1[-2,]),1-sum(x1[-3,3])/sum(x1[-3,]),1-sum(x1[-4,4])/sum(x1[-4,]))
sensitivity<-rbind(sensitivity,sen)
specificity<-rbind(specificity,spe)
}
x<-x/100
x1<-x
x1[,3]<-x1[,3]+x1[,4]
x1<-x1[,-4]
x1[3,]<-x1[3,]+x1[4,]
x1<-x1[-4,]
SEN<-c(ci95(sensitivity[,1]),ci95(sensitivity[,2]),ci95(sensitivity[,3]),ci95(sensitivity[,4]))
SPE<-c(ci95(specificity[,1]),ci95(specificity[,2]),ci95(specificity[,3]),ci95(specificity[,4]))
rlt<-cbind(x1[,1:4],SEN,SPE)
rf.importance.batch1<-rf.importance
write.table(rlt,file="random.forest.confusion.batch1.RRBS.txt",sep="\t",quote=F,col.names=NA,row.names=T)
write.table(rf.importance.batch1,file="random.forest.confusion.batch1.feature.importance.RRBS.txt",sep="\t",quote=F,col.names=NA,row.names=T)

newdata<-newdata1[match(cor,rownames(newdata1)),]
dim(newdata)
colnames(newdata)
data<-newdata
# bilud the prediction model
# batch 1
# batch 2
ob4<-grep("RRBS-6P",colnames(data))
ob5<-grep("RRBS-7P",colnames(data))

pred1<-data[,c(ob4,ob5)]
head(pred1)
lab1<-as.factor(c(rep("Colon",length(ob4)),rep("Lung",length(ob5))))
colnames(pred1)<-lab1
hgsi1<-gsi(data.matrix(pred1))
pred2<-pred1[match(subset(hgsi1,GSI>0.3)[,1],rownames(pred1)),]
batch2_feature<-pred2
dim(pred2)
head(pred2)
lab2<-lab1
dim(pred2)

library("randomForest")
x<-matrix(0,ncol=3,nrow=2)
rf.importance<-rep(0,nrow(pred2))
bestmtry<-tuneRF(x=t(pred2),y=lab2,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
for(i in 1:100){
rf.model1<-randomForest(x=t(pred2),y=lab2,mtry=best.mtry)
x<-rf.model1$confusion+x
rf.importance<-rf.importance+rf.model1$importance/100
x1<-x
sen<-c(x1[1,1]/sum(x1[1,]),x1[2,2]/sum(x1[2,]))
spe<-c(1-sum(x1[-1,1])/sum(x1[-1,]),1-sum(x1[-2,2])/sum(x1[-2,]))
sensitivity<-rbind(sensitivity,sen)
specificity<-rbind(specificity,spe)
}
x<-x/100
x1<-x
SEN<-c(ci95(sensitivity[,1]),ci95(sensitivity[,2]))
SPE<-c(ci95(specificity[,1]),ci95(specificity[,2]))
rlt<-cbind(x1[,1:2],SEN,SPE)
rf.importance.batch2<-rf.importance
write.table(rlt,file="random.forest.confusion.batch2.RRBS.txt",sep="\t",quote=F,col.names=NA,row.names=T)
write.table(rf.importance.batch2,file="random.forest.confusion.batch2.feature.importance.RRBS.txt",sep="\t",quote=F,col.names=NA,row.names=T)

# feature comparsion
match(rownames(rf.importance.batch1), rownames(rf.importance.batch2))

# batch 1 and batch 2 together
data<-newdata1
ob1<-grep("6-P",colnames(data))
ob2<-grep("7-P",colnames(data))
# ob3<-grep("PC-P",colnames(data))
ob4<-grep("RRBS-6P",colnames(data))
ob5<-grep("RRBS-7P",colnames(data))
ob6<-grep("NC-P",colnames(data))

pred1<-data[,c(ob1,ob2,ob4,ob5,ob6)]
head(pred1)
lab1<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Colon",length(ob4)),rep("Lung",length(ob5)),rep("Normal",length(ob6))))
colnames(pred1)<-lab1
hgsi1<-gsi(data.matrix(pred1))
pred2<-pred1[match(subset(hgsi1,GSI>0.44)[,1],rownames(pred1)),]
batch12_feature<-pred2
write.table(batch12_feature,file="random.forest.confusion.batch1.and.2.RRBS.InputMatrix.txt",sep="\t",quote=F,col.names=NA,row.names=T)
head(pred2)
lab2<-as.factor(c(rep("Colon",length(ob1)),rep("Lung",length(ob2)),rep("Colon",length(ob4)),rep("Lung",length(ob5)),rep("Normal",length(ob6))))
dim(pred2)

library("randomForest")
x<-matrix(0,ncol=4,nrow=3)
rf.importance<-rep(0,nrow(pred2))
bestmtry<-tuneRF(x=t(pred2),y=lab2,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
for(i in 1:100){
rf.model1<-randomForest(x=t(pred2),y=lab2,mtry=best.mtry)
x<-rf.model1$confusion+x
rf.importance<-rf.importance+rf.model1$importance/100
x1<-x
sen<-c(x1[1,1]/sum(x1[1,]),x1[2,2]/sum(x1[2,]),x1[3,3]/sum(x1[3,]))
spe<-c(1-sum(x1[-1,1])/sum(x1[-1,]),1-sum(x1[-2,2])/sum(x1[-2,]),1-sum(x1[-3,3])/sum(x1[-3,]))
sensitivity<-rbind(sensitivity,sen)
specificity<-rbind(specificity,spe)
}
x<-x/100
x1<-x
SEN<-c(ci95(sensitivity[,1]),ci95(sensitivity[,2]),ci95(sensitivity[,3]))
SPE<-c(ci95(specificity[,1]),ci95(specificity[,2]),ci95(specificity[,3]))
rlt<-cbind(x1[,1:3],SEN,SPE)
rf.importance.batch12<-rf.importance
write.table(rlt,file="random.forest.confusion.batch12.RRBS.txt",sep="\t",quote=F,col.names=NA,row.names=T)
write.table(rf.importance.batch12,file="random.forest.confusion.batch12.feature.importance.RRBS.txt",sep="\t",quote=F,col.names=NA,row.names=T)
```

Now, we need to identify some cancer biomarker through comparion between tissue and normal. 
```{r}
setwd("/home/shg047/monod/rrbs_kun")
setwd("/home/sguo/monod/rrbs_kun")
newdata1<-read.table("rrbs.kun.mhl.na.impute.txt",head=T,as.is=T,row.names=1,check.names=F)
bed2<-cor2bed(rownames(newdata1))
dim(newdata1)
dim(bed2)
head(newdata1)
data<-newdata1
ob1<-grep("6-P",colnames(data))
ob2<-grep("7-P",colnames(data))
ob3<-grep("PC-P",colnames(data))
ob4<-grep("RRBS-6P",colnames(data))
ob5<-grep("RRBS-7P",colnames(data))
ob6<-grep("NC-P",colnames(data))

length(ob1)+length(ob4)
data_colon<-data[,c(ob1,ob4,ob6)]
data_lung<-data[,c(ob2,ob5,ob6)]
data_pancrease<-data[,c(ob3,ob6)]

p1<-ttestPValue(data_colon,1:30,31:50)
p1.fdr<-p.adjust(p2[,2],"fdr")
sum(p1.fdr<0.05)
pred2=data_colon[which(p1.fdr<0.05),]
lab2<-c(rep("lung",30),rep("normal",20))
bestmtry<-tuneRF(x=t(pred2),y=lab2,ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
rf.model1<-randomForest(x=t(pred2),y=lab2,mtry=best.mtry)
x<-rf.model1$confusion+x

write.table(p1,file="colon.dmr.fdr0.5.txt",col.names=NA,row.names=T,sep="\t")

p2<-ttestPValue(data_lung,1:29,30:49)
p2.fdr<-p.adjust(p2[,2],"fdr")
sum(p2.fdr<0.05)
write.table(p2,file="lung.dmr.fdr0.5.txt",col.names=NA,row.names=T,sep="\t")

p3<-ttestPValue(data_pancrease,1:10,11:30)
p3.fdr<-p.adjust(p3[,2],"fdr")
sum(p3.fdr<0.05)
write.table(p3,file="pancrease.dmr.fdr0.5.txt",col.names=NA,row.names=T,sep="\t")

```

Dr. Zhang still want to get full prediction model from WGBS and then test them in RRBS. Now we try to decrease the tissue numbers and try to balance the sample size for different tissies.

He suggest to increase the sample size from RRBS ENCODE. I don't think it is a good idea since more data from different group would increase the heterogeneity of the data.

1, I am wondering whether I can make the prediction with tradtional methylation level rather than MHL
2, Can we cluster sample together with average methylation level

```{r}
setwd("/home/shg047/work/monod/rrbs_kun")
data<-read.table("rrbs.kun.mhl.txt",head=T,sep="\t",as.is=T,row.names=1,check.names=F)
newdata<-RawNARemove(data,missratio=0.45)
newdata<-ColNARemove(newdata,missratio=0.5)
dim(data)
dim(newdata)
newdata<-impute.knn(data.matrix(newdata))$data
colnames(newdata)
length(grep("6-P",colnames(newdata)))
length(grep("7-P",colnames(newdata)))
length(grep("PC-P",colnames(newdata)))
length(grep("NC-P",colnames(newdata)))
length(grep("6-T",colnames(newdata)))
length(grep("7-T",colnames(newdata)))
length(grep("PC-T",colnames(newdata)))
length(grep("RRBS-6P",colnames(newdata)))
length(grep("RRBS-7P",colnames(newdata)))
# all kinds of  biomarker and tissue specifal regions
```

Update: Jan-19-2015
Merge Haib and WGBS data


```{r}
# load all the pre-functions advance
# read sample information

saminfo1<-read.table("/home/shg047/work/monod/hap/wgbs/All_chromosomes_combined/newsaminfo.txt",head=T,sep="\t",as.is=T)
saminfo2<-read.table("/home/shg047/work/monod/2016/sampleinfo.txt",head=F,sep="\t",as.is=T)
saminfo3<-read.table("/home/shg047/work/monod/2016/haib.select.sample.txt",head=F,sep="\t",as.is=T)

## WGBS dataset collection
setwd("/home/shg047/work/monod/hap/wgbs/All_chromosomes_combined")
newdata2<-read.table("wgbs.mhl.txt",head=T,sep="\t",as.is=T,row.names=1,check.names=F)
file1<-newdata2
f2<-RawNARemove(file1,missratio=0.3)
f2<-impute.knn(data.matrix(f2))$data
colnames(f2) 
colnames(f2)<-gsub("_","-",colnames(f2))
colname2<-unlist(lapply(colnames(f2),function(x) unlist(strsplit(x,"[.]"))[1]))
colname2
colnames(f2)<-colname2
saminfo2<-read.table("/home/shg047/work/monod/hap/wgbs/All_chromosomes_combined/newsaminfo.txt",head=T,sep="\t",as.is=T)

saminfo2<-saminfo2[match(colname2,saminfo2[,1]),]
saminfo2
colnames(f2)<-saminfo2[,2]
colnames(f2)
newdata2<-f2
incl<-grep("Brain|Colon|Intestine|Kidney|Liver|Lung|Pancreas|Spleen|Stomach|WB",colnames(newdata2))
gsidata<-newdata2[,incl]
newdata1<-gsidata

## Haib Dataset
haib<-read.table("/home/shg047/2016/methHapLoad.matrix.txt",head=T,row.names=1)
f2<-RawNARemove(haib,missratio=0.3)
f2<-impute.knn(data.matrix(f2))$data
colnames(f2) 
newdata2<-f2
newdata2<-newdata2[,! is.na(match(unlist(strsplit(colnames(newdata2)[grep("trim",colnames(newdata2))],"_"))[seq(1,78,by=2)],saminfo3[,1]))]
colnames(newdata2)<-saminfo3[na.omit(match(unlist(strsplit(colnames(newdata2)[grep("trim",colnames(newdata2))],"_"))[seq(1,78,by=2)],saminfo3[,1])),3]

## rrbs dataset
data<-read.table("/home/shg047/work/monod/rrbs_kun/rrbs.kun.mhl.txt",head=T,sep="\t",as.is=T,row.names=1,check.names=F)
incl<-c(grep("6-P|7-P|PC-P|NC-P|RRBS-6P|RRBS-7P",colnames(data)))
data<-data[,incl]
newdata<-RawNARemove(data,missratio=0.45)
newdata<-ColNARemove(newdata,missratio=0.5)
dim(data)
dim(newdata)
newdata<-impute.knn(data.matrix(newdata))$data
colnames(newdata)

length(grep("6-P",colnames(newdata)))
length(grep("7-P",colnames(newdata)))
length(grep("PC-P",colnames(newdata)))
length(grep("NC-P",colnames(newdata)))
length(grep("6-T",colnames(newdata)))
length(grep("7-T",colnames(newdata)))
length(grep("PC-T",colnames(newdata)))
length(grep("RRBS-6P",colnames(newdata)))
length(grep("RRBS-7P",colnames(newdata)))

x1<-(grep("6-P",colnames(newdata)))
x2<-(grep("7-P",colnames(newdata)))
x3<-(grep("PC-P",colnames(newdata)))
x4<-(grep("NC-P",colnames(newdata)))
x5<-(grep("6-T",colnames(newdata)))
x6<-(grep("7-T",colnames(newdata)))
x7<-(grep("PC-T",colnames(newdata)))
x8<-(grep("RRBS-6P",colnames(newdata)))
x9<-(grep("RRBS-7P",colnames(newdata)))

newdata3<-newdata[,c(x1,x2,x3,x4,x8,x9)]
colnames(newdata3)[grep("6P|6-P",colnames(newdata3))]<-"Colon"
colnames(newdata3)[grep("7P|7-P",colnames(newdata3))]<-"Lung"
colnames(newdata3)[grep("PC-P",colnames(newdata3))]<-"Pancreas"
colnames(newdata3)[grep("NC-P",colnames(newdata3))]<-"WB"

colnames(newdata3)

# merge wgbs and Haib
bed1<-cor2bed(rownames(newdata1))
bed2<-cor2bed(rownames(newdata2))
bed3<-cor2bed(rownames(newdata3))
bed21<-Rbedtools("intersectBed",bed2,bed1,opt.string="-wa -u")
bed321<-Rbedtools("intersectBed",bed3,bed21,opt.string="-wa -u")
dim(bed21)
dim(bed321)
cor<-bed2cor(bed321)


# built prediction model
dim(newdata1)
dim(newdata2)
dim(newdata3)

wgbs_new<-newdata1[match(cor,rownames(newdata1)),]
haib_new<-newdata2[match(cor,rownames(newdata2)),]
kun_new<-newdata3[match(cor,rownames(newdata3)),]

head(wgbs_new)
head(rrbs_new)
dim(wgbs_new)
dim(haib_new)
dim(kun_new)

data1<-cbind(wgbs_new,haib_new,kun_new)
hgsi1<-gsi(data1)
hgsi_sub1<-subset(hgsi1,GSI>0.01)
dim(hgsi_sub)
table(hgsi_sub1[,2])

regionsele<-c()
for(i in 1:10){
  nd<-subset(hgsi_sub,group==names(table(hgsi_sub1[,2]))[i])
  if(nrow(nd)>5){
    regionsele<-rbind(regionsele,nd[order(nd[,3],decreasing=T)[1:5],])
  }else{
    regionsele<-rbind(regionsele,nd) 
  }
}

bed<-cor2bed(regionsele[,1])
new<-data.frame(bed,regionsele)

write.table(new,file="predictor.selected.regions.txt",sep="\t",col.names=F,row.names=F,quote=F)
head(new)
anno<-read.table("gene.annotation.txt",sep="\t",as.is=T,head=T)


# for gsi
data2<-cbind(wgbs_new,haib_new,kun_new)
dim(data2)
data2<-na.omit(data2)
dim(data2)
write.table(data2,file="wgbs.haib.kun.merge.raw.data.txt",col.names=NA,row.names=T,sep="\t")
data2<-read.table("wgbs.haib.merge.raw.data.txt",row.names=1,head=T,sep="\t",as.is=T,check.names=F)
dim(data2)
hgsi<-gsi(data=data2)
write.table(hgsi,file="wgbs.haib.kun.merge.gsi.rlt.txt",col.names=NA,row.names=T,sep="\t")
hgsi<-read.table("wgbs.haib.kun.merge.gsi.rlt.txt",row.names=1,head=T,sep="\t",as.is=T,check.names=F)
hgsi_sub<-subset(hgsi,GSI>0.01)
dim(hgsi_sub)
table(hgsi_sub[,2])
subset(group==Pancreas,data=hgsi)


# hgsi corresponding genes
bed3<-cor2bed(hgsi_sub[,1])
cor3<-bed2cor(bed3)
write.table(cor3,file="high.gsi.txt",sep="\t",quote=F,row.names=F,col.names=F)
gsi_gene<-read.table("/home/shg047/work/monod/2016/gsi.gene.txt",sep="\t")
# random forest prediction based on high gsi regions
library("randomForest")
data2new<-data2[match(hgsi_sub[,1],rownames(data2)),]
train=data2new[,1:(ncol(wgbs_new)+ncol(haib_new))]
test=data2new[,(ncol(wgbs_new)+ncol(haib_new)+1):ncol(data2)]

as.factor(colnames(train))
as.factor(colnames(test))

bestmtry<-tuneRF(x=t((train)),y=(as.factor(colnames(train))),ntreeTry=1000, stepFactor=2, improve=0.05,trace=TRUE, plot=TRUE, doBest=T)
best.mtry=bestmtry$mtry
best.mtry
rf.model1<-randomForest(x=t(train),y=as.factor(colnames(train)),mtry=best.mtry)

hgsi_test<-gsi(test)
newtest<-test[match(subset(hgsi_test,GSI>0.1)[,1],rownames(test)),]
rf.model2<-randomForest(x=t(newtest),y=as.factor(colnames(newtest)))
rf.model2


bed21<-Rbedtools("intersectBed",gsi_gene[,1:3],cor2bed(as.character(hgsi_sub[,1])),opt.string="-wa -u")
match(bed2cor(bed21)
      # add bio-info
      # scp sampleinfo.txt shg047@genome-miner.ucsd.edu:/home/shg047/work/monod/2016/
      saminfo1<-read.table("/home/shg047/work/monod/hap/wgbs/All_chromosomes_combined/newsaminfo.txt",head=T,sep="\t",as.is=T)
      saminfo2<-read.table("/home/shg047/work/monod/2016/sampleinfo.txt",head=F,sep="\t",as.is=T)
      saminfo3<-read.table("/home/shg047/work/monod/2016/haib.select.sample.txt",head=F,sep="\t",as.is=T)
      
      hgsi<-gsi(gsidata)
      hgsi_sub<-subset(hgsi,GSI>0.5)
      bed3<-cor2bed(hgsi_sub[,1])
      dim(gsidata)
      dim(hgsi_sub)
      
      ```
      
      identify different genes
      
      ```{r}
      setwd("/media/LTS_33T/SG_LTS33T/monod/hap/wgbs/All_chromosomes_combined")
      data<-read.table("/home/shg047/work/monod/rrbs_kun/rrbs.kun.mhl.txt",head=T,sep="\t",as.is=T,row.names=1,check.names=F)
      newdata<-RawNARemove(data,missratio=0.45)
      newdata<-ColNARemove(newdata,missratio=0.5)
      dim(data)
      dim(newdata)
      newdata<-impute.knn(data.matrix(newdata))$data
      colnames(newdata)
      
      length(grep("6-P",colnames(newdata)))
      length(grep("7-P",colnames(newdata)))
      length(grep("PC-P",colnames(newdata)))
      length(grep("NC-P",colnames(newdata)))
      length(grep("6-T",colnames(newdata)))
      length(grep("7-T",colnames(newdata)))
      length(grep("PC-T",colnames(newdata)))
      length(grep("RRBS-6P",colnames(newdata)))
      length(grep("RRBS-7P",colnames(newdata)))
      
      x1<-(grep("6-P",colnames(newdata)))
      x2<-(grep("7-P",colnames(newdata)))
      x3<-(grep("PC-P",colnames(newdata)))
      x4<-(grep("NC-P",colnames(newdata)))
      x5<-(grep("6-T",colnames(newdata)))
      x6<-(grep("7-T",colnames(newdata)))
      x7<-(grep("PC-T",colnames(newdata)))
      x8<-(grep("RRBS-6P",colnames(newdata)))
      x9<-(grep("RRBS-7P",colnames(newdata)))
      
      
      newdata1<-newdata[,c(x1,x8,x4)]
      newdata2<-newdata[,c(x2,x9,x4)]
      newdata3<-newdata[,c(x3,x4)]
      
      newdata4<-newdata[,c(x1,x5)]
      newdata5<-newdata[,c(x2,x6)]
      
      
      rlt1<-ttestPValue(newdata1,1:(length(x1)+length(x8)),(length(x1)+length(x8)+1):ncol(newdata1))
      rlt2<-ttestPValue(newdata1,1:(length(x2)+length(x9)),(length(x2)+length(x9)+1):ncol(newdata2))
      rlt3<-ttestPValue(newdata1,1:(length(x3)),(length(x3)+1):ncol(newdata3))
      
      rltt1<-subset(rlt1,FDR<0.05)
      rltt2<-subset(rlt2,FDR<0.05)
      rltt3<-subset(rlt3,FDR<0.05)
      
      write.table(rltt1,file="DMMHL-6P.txt",col.names=NA,row.names=T,sep="\t",quote=F)
      write.table(rltt2,file="DMMHL-7P.txt",col.names=NA,row.names=T,sep="\t",quote=F)
      write.table(rltt3,file="DMMHL-PCP.txt",col.names=NA,row.names=T,sep="\t",quote=F)
      
      newdata3<-newdata[,c(x1,x2,x3,x4,x8,x9)]
      
      colnames(newdata3)[grep("6P|6-P",colnames(newdata3))]<-"Colon"
      colnames(newdata3)[grep("7P|7-P",colnames(newdata3))]<-"Lung"
      colnames(newdata3)[grep("PC-P",colnames(newdata3))]<-"Pancreas"
      colnames(newdata3)[grep("NC-P",colnames(newdata3))]<-"WB"
      colnames(newdata3)
      
      ```
      
      
      
      
      
      
      
      
      